# 筛选apparent temperature > 33 ℃的数据

In [ ]:
import glob
import os
import pandas as pd
input_dir = r'D:\seoul\a_airtem'
csv_file = os.path.join(input_dir, 'AT_2015_2024.csv')
df = pd.read_csv(csv_file, skiprows=3, encoding='euc_kr')

df

In [ ]:
print(df)

In [ ]:
import pandas as pd

# 1. 读取并解析日期
path = r'D:\seoul\a_airtem\AT_2015_2024.csv'
df = pd.read_csv(path, skiprows=3, encoding='euc_kr', parse_dates=['일자'])

# 2. 按日期排序
df = df.sort_values('일자')
print(df)
# 3. 标记“体感温度 > 33”
df['is_hot'] = df['체감온도(°C)'] > 33

# 4. 计算前一天记录与当前行的日期差
df['prev_diff'] = df['일자'] - df['일자'].shift(1)

# 5. 标记“与前一天连续且两天都 >33”
df['two_day_consec'] = (
    df['is_hot'] &
    df['is_hot'].shift(1) &
    (df['prev_diff'] == pd.Timedelta(days=1))
)

# 6. 把“连续两天”中的两个日期都选出来
mask = df['two_day_consec'] | df['two_day_consec'].shift(-1).fillna(False)
result = df.loc[mask, ['일자', '기온(°C)', '습도(%rh)', '체감온도(°C)']]

# 7. 输出
print("=== 连续两天体感温度 > 33°C 的记录 ===")
print(result.to_string(index=False))
result.to_csv(r'D:\seoul\a_airtem\result\highest_temps_over_33.csv')

In [9]:
import pandas as pd

path = r'D:\seoul\a_airtem\result\highest_temps_over_33.csv'

# 用和写出时一样的编码来读取
df = pd.read_csv(path, encoding='utf-8-sig', parse_dates=['일자'])

# 现在就能打印 일자 列了
print(df['일자'])

0     2015-07-10
1     2015-07-11
2     2015-07-30
3     2015-07-31
4     2015-08-06
         ...    
158   2024-09-10
159   2024-09-11
160   2024-09-17
161   2024-09-18
162   2024-09-19
Name: 일자, Length: 163, dtype: datetime64[ns]


In [11]:
import numpy as np

unique_dates = np.array(df['일자'].dt.strftime('%Y-%m-%d'))
print(unique_dates)

['2015-07-10' '2015-07-11' '2015-07-30' '2015-07-31' '2015-08-06'
 '2015-08-07' '2015-08-08' '2016-07-30' '2016-07-31' '2016-08-01'
 '2016-08-03' '2016-08-04' '2016-08-05' '2016-08-06' '2016-08-07'
 '2016-08-08' '2016-08-09' '2016-08-10' '2016-08-11' '2016-08-12'
 '2016-08-13' '2016-08-14' '2016-08-15' '2016-08-16' '2016-08-17'
 '2016-08-18' '2016-08-19' '2016-08-20' '2016-08-21' '2016-08-22'
 '2016-08-23' '2017-07-19' '2017-07-20' '2017-07-21' '2017-07-22'
 '2017-08-01' '2017-08-02' '2017-08-03' '2017-08-04' '2017-08-05'
 '2017-08-06' '2017-08-07' '2018-07-15' '2018-07-16' '2018-07-19'
 '2018-07-20' '2018-07-21' '2018-07-22' '2018-07-23' '2018-07-24'
 '2018-07-25' '2018-07-26' '2018-07-27' '2018-07-28' '2018-07-29'
 '2018-07-30' '2018-07-31' '2018-08-01' '2018-08-02' '2018-08-03'
 '2018-08-04' '2018-08-05' '2018-08-06' '2018-08-07' '2018-08-08'
 '2018-08-09' '2018-08-10' '2018-08-11' '2018-08-12' '2018-08-13'
 '2018-08-14' '2018-08-15' '2018-08-16' '2019-08-02' '2019-08-03'
 '2019-08-

# 02 需要下载全部的对应文件的 B10 和 Qixel 处理 extreme heat date
 https://earthexplorer.usgs.gov/

In [13]:
landsat_8 = r'D:\seoul\b_satellite_img\landsat_ot_c2_l2.csv'
df = pd.read_csv(landsat_8, encoding='ISO-8859-1')
df

,Landsat Product Identifier L2,Landsat Product Identifier L1,Landsat Scene Identifier,Date Acquired,Collection Category,Collection Number,WRS Path,WRS Row,Target WRS Path,Target WRS Row,...,Corner Upper Left Latitude,Corner Upper Left Longitude,Corner Upper Right Latitude,Corner Upper Right Longitude,Corner Lower Left Latitude,Corner Lower Left Longitude,Corner Lower Right Latitude,Corner Lower Right Longitude,Display ID,Entity ID
0,LC08_L2SP_116034_20150501_20200909_02_T1,LC08_L1TP_116034_20150501_20200909_02_T1,LC81160342015121LGN01,2015/05/01,T1,2,116,34,116,34,...,38.51088,125.21327,38.56704,127.90608,36.36027,125.32025,36.41226,127.93707,LC08_L2SP_116034_20150501_20200909_02_T1,LC81160342015121LGN01
1,LC08_L2SP_115034_20150510_20200909_02_T1,LC08_L1TP_115034_20150510_20200909_02_T1,LC81150342015130LGN01,2015/05/10,T1,2,115,34,115,34,...,38.51915,126.80568,38.53876,129.47274,36.38712,126.86713,36.40528,129.45949,LC08_L2SP_115034_20150510_20200909_02_T1,LC81150342015130LGN01
2,LC08_L2SP_115035_20150510_20200909_02_T1,LC08_L1TP_115035_20150510_20200909_02_T1,LC81150352015130LGN01,2015/05/10,T1,2,115,35,115,35,...,37.08357,126.40232,37.11203,129.02139,34.95164,126.47141,34.97797,129.02082,LC08_L2SP_115035_20150510_20200909_02_T1,LC81150352015130LGN01
3,LC08_L2SP_116034_20150517_20200909_02_T1,LC08_L1TP_116034_20150517_20200909_02_T1,LC81160342015137LGN01,2015/05/17,T1,2,116,34,116,34,...,38.51253,125.26477,38.56755,127.96117,36.36180,125.37031,36.41272,127.99060,LC08_L2SP_116034_20150517_20200909_02_T1,LC81160342015137LGN01
4,LC08_L2SP_115034_20150526_20200909_02_T1,LC08_L1TP_115034_20150526_20200909_02_T1,LC81150342015146LGN01,2015/05/26,T1,2,115,34,115,34,...,38.52010,126.85726,38.53852,129.52782,36.38801,126.91727,36.40507,129.51302,LC08_L2SP_115034_20150526_20200909_02_T1,LC81150342015146LGN01
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
356,LC09_L2SP_115035_20240915_20240916_02_T1,LC09_L1TP_115035_20240915_20240915_02_T1,LC91150352024259LGN00,2024/09/15,T1,2,115,35,115,35,...,37.08408,126.42592,37.11202,129.04840,34.95212,126.49438,34.97796,129.04711,LC09_L2SP_115035_20240915_20240916_02_T1,LC91150352024259LGN00
357,LC09_L2SP_116034_20240922_20240924_02_T1,LC09_L1TP_116034_20240922_20240922_02_T1,LC91160342024266LGN00,2024/09/22,T1,2,116,34,116,34,...,38.51016,125.27521,38.56493,127.97154,36.36211,125.38032,36.41280,128.00063,LC09_L2SP_116034_20240922_20240924_02_T1,LC91160342024266LGN00
358,LC08_L2SP_115034_20240923_20240928_02_T1,LC08_L1TP_115034_20240923_20240928_02_T1,LC81150342024267LGN00,2024/09/23,T1,2,115,34,115,34,...,38.51979,126.84007,38.53860,129.51060,36.38771,126.90056,36.40514,129.49629,LC08_L2SP_115034_20240923_20240928_02_T1,LC81150342024267LGN00
359,LC08_L2SP_115035_20240923_20240928_02_T1,LC08_L1TP_115035_20240923_20240928_02_T1,LC81150352024267LGN00,2024/09/23,T1,2,115,35,115,35,...,37.08437,126.43941,37.11201,129.05853,34.95239,126.50751,34.97796,129.05697,LC08_L2SP_115035_20240923_20240928_02_T1,LC81150352024267LGN00


In [15]:
# 过滤只保留 Collection Category 为 T1 的数据
df_t1 = df[df['Collection Category'] == 'T1']
df_t1['Date Acquired']

0      2015/05/01
1      2015/05/10
2      2015/05/10
3      2015/05/17
4      2015/05/26
          ...    
356    2024/09/15
357    2024/09/22
358    2024/09/23
359    2024/09/23
360    2024/09/30
Name: Date Acquired, Length: 298, dtype: object

In [16]:
unique_dates = unique_dates
acquired    = df_t1['Date Acquired'].to_numpy()
#print(unique_dates)
#print(acquired)
# 打印类型和元素的数据类型（如果有 dtype 属性）
print("unique_dates →", type(unique_dates), getattr(unique_dates, 'dtype', None))
print("acquired    →", type(acquired), acquired.dtype)

unique_dates → <class 'numpy.ndarray'> object
acquired    → <class 'numpy.ndarray'> object


In [18]:
import pandas as pd

unique_dates = unique_dates
acquired    = pd.to_datetime(
    df_t1['Date Acquired'],
    format='%Y/%m/%d',  # 注意这里是“/”
    errors='coerce'
).values
unique_dates_clean = unique_dates.astype('datetime64[D]')
acquired_clean      = acquired.astype   ('datetime64[D]')

In [19]:
import pandas as pd

# 求交集
common = set(acquired_clean) & set(unique_dates_clean)
# 把 set 转成列表并排序
common_sorted = sorted(common)
# （如果需要字符串格式，可以再 map 一次 str）
common_str = [str(d) for d in common_sorted]
print(f"共有 {len(common_str)} 个相同的日期：{common_str}")
for d in common_str:
    print(d)

共有 21 个相同的日期：['2016-07-31', '2016-08-07', '2016-08-16', '2016-08-23', '2017-08-03', '2018-07-21', '2018-07-28', '2018-08-06', '2018-08-13', '2019-08-09', '2020-08-18', '2021-07-29', '2021-08-05', '2023-07-27', '2023-08-03', '2023-08-04', '2023-08-19', '2024-08-05', '2024-08-06', '2024-08-13', '2024-08-14']
2016-07-31
2016-08-07
2016-08-16
2016-08-23
2017-08-03
2018-07-21
2018-07-28
2018-08-06
2018-08-13
2019-08-09
2020-08-18
2021-07-29
2021-08-05
2023-07-27
2023-08-03
2023-08-04
2023-08-19
2024-08-05
2024-08-06
2024-08-13
2024-08-14


In [20]:
sorted_dates = [str(d).replace("-", "") for d in common_sorted]
print(sorted_dates)  # 输出: 20240618

['20160731', '20160807', '20160816', '20160823', '20170803', '20180721', '20180728', '20180806', '20180813', '20190809', '20200818', '20210729', '20210805', '20230727', '20230803', '20230804', '20230819', '20240805', '20240806', '20240813', '20240814']


# 然后下载对应的文件

In [21]:
import os
import rasterio
import geopandas as gpd
from rasterio.mask import mask
from glob import glob

def clip_rasters_with_shapefile(raster_folder, shapefile_path, output_folder):
    # Ensure the output directory exists
    os.makedirs(output_folder, exist_ok=True)

    # Load the shapefile as a GeoDataFrame
    nyc_shape = gpd.read_file(shapefile_path)

    # Loop through each raster in the raster folder
    for raster_path in glob(os.path.join(raster_folder, '*.tif')):
        with rasterio.open(raster_path) as src:
            raster_crs = src.crs

            # Ensure the shapefile is in the same CRS as the raster
            if nyc_shape.crs != raster_crs:
                nyc_shape = nyc_shape.to_crs(raster_crs)

            # Clip the raster using the shapefile
            out_image, out_transform = mask(src, nyc_shape.geometry, crop=True)

            # Update metadata for the output raster
            out_meta = src.meta.copy()
            out_meta.update({
                "driver": "GTiff",
                "height": out_image.shape[1],
                "width": out_image.shape[2],
                "transform": out_transform,
                "crs": raster_crs  # Use the raster's CRS for the output
            })

            # Save the clipped raster to the output folder
            output_raster_path = os.path.join(output_folder, f"clipped_{os.path.basename(raster_path)}")
            if os.path.isdir(output_raster_path):
                continue
            with rasterio.open(output_raster_path, "w", **out_meta) as dest:
                dest.write(out_image)

            print(f"Clipped raster saved to: {output_raster_path}")

In [22]:
# Paths to the folder and shapefile
raster_folder = r'D:\seoul\b_satellite_img\extreme\original'
seoul_boundary_file = r'D:\seoul\Final_data\Admin_boundary\Seoul_boundary.shp'
shp_seoul = seoul_boundary_file
output_folder = r'D:\seoul\b_satellite_img\extreme\clipped'
os.makedirs(output_folder, exist_ok=True)

clip_rasters_with_shapefile(raster_folder, shp_seoul, output_folder)

Clipped raster saved to: D:\seoul\b_satellite_img\extreme\clipped\clipped_LC08_L2SP_116034_20160807_20200906_02_T1_QA_PIXEL.TIF
Clipped raster saved to: D:\seoul\b_satellite_img\extreme\clipped\clipped_LC08_L2SP_116034_20160807_20200906_02_T1_SR_B4.TIF
Clipped raster saved to: D:\seoul\b_satellite_img\extreme\clipped\clipped_LC08_L2SP_116034_20160807_20200906_02_T1_SR_B5.TIF
Clipped raster saved to: D:\seoul\b_satellite_img\extreme\clipped\clipped_LC08_L2SP_116034_20160807_20200906_02_T1_ST_B10.TIF
Clipped raster saved to: D:\seoul\b_satellite_img\extreme\clipped\clipped_LC08_L2SP_116034_20160823_20200906_02_T1_QA_PIXEL.TIF
Clipped raster saved to: D:\seoul\b_satellite_img\extreme\clipped\clipped_LC08_L2SP_116034_20160823_20200906_02_T1_SR_B4.TIF
Clipped raster saved to: D:\seoul\b_satellite_img\extreme\clipped\clipped_LC08_L2SP_116034_20160823_20200906_02_T1_SR_B5.TIF
Clipped raster saved to: D:\seoul\b_satellite_img\extreme\clipped\clipped_LC08_L2SP_116034_20160823_20200906_02_T1_ST_

In [23]:
import os
import rasterio
import numpy as np
import pandas as pd
import glob
# Mapping from month names to numbers for date comparison
month_to_num = {
    'Jan': '01', 'Feb': '02', 'Mar': '03', 'Apr': '04', 'May': '05', 'Jun': '06',
    'Jul': '07', 'Aug': '08', 'Sep': '09', 'Oct': '10', 'Nov': '11', 'Dec': '12'
}

def calculate_pixel_proportion(sorted_dates, base_folder, pixel_values=[21824, 21952]):
    for sorted_data in sorted_dates:
        pattern = r'{}\*_*{}_*_QA_PIXEL.TIF'.format(base_folder, sorted_data)
        matched_files = glob.glob(pattern)
        if not matched_files:
            continue
        with rasterio.open(matched_files[0]) as src:
            qa_pixel_data = src.read(1)
        non_zero_pixels = qa_pixel_data[qa_pixel_data > 0]
        total_non_zero_pixels = non_zero_pixels.size
        unique, counts = np.unique(non_zero_pixels, return_counts=True)
        pixel_counts = dict(zip(unique, counts))
        pixel_proportions = {value: count / total_non_zero_pixels for value, count in pixel_counts.items()}

        print(f"📅 {sorted_data}:")
        val_all = 0
        for val in pixel_values:
            prop = pixel_proportions.get(val, 0)
            print(f"  - 像素值 {val} 的占比: {prop:.2%}")
            val_all = val_all + prop
        print(val_all)


# Example usage
base_folder = r'D:\seoul\b_satellite_img\extreme\clipped'

calculate_pixel_proportion(sorted_dates, base_folder)

📅 20160807:
  - 像素值 21824 的占比: 78.39%
  - 像素值 21952 的占比: 4.13%
0.8252899697584389
📅 20160823:
  - 像素值 21824 的占比: 19.05%
  - 像素值 21952 的占比: 0.12%
0.1916066665180595
📅 20180813:
  - 像素值 21824 的占比: 1.81%
  - 像素值 21952 的占比: 0.00%
0.018127103720380733
📅 20200818:
  - 像素值 21824 的占比: 11.17%
  - 像素值 21952 的占比: 0.24%
0.1141184250611147
📅 20210805:
  - 像素值 21824 的占比: 14.41%
  - 像素值 21952 的占比: 0.65%
0.15064606971162778
📅 20230803:
  - 像素值 21824 的占比: 30.11%
  - 像素值 21952 的占比: 0.59%
0.3069243515154217
📅 20230804:
  - 像素值 21824 的占比: 14.74%
  - 像素值 21952 的占比: 0.80%
0.1553925830156855
📅 20230819:
  - 像素值 21824 的占比: 68.86%
  - 像素值 21952 的占比: 3.80%
0.7265583320330204
📅 20240805:
  - 像素值 21824 的占比: 8.68%
  - 像素值 21952 的占比: 0.25%
0.08935749686067333
📅 20240806:
  - 像素值 21824 的占比: 8.68%
  - 像素值 21952 的占比: 0.25%
0.08935749686067333
📅 20240813:
  - 像素值 21824 的占比: 59.70%
  - 像素值 21952 的占比: 3.25%
0.6295118997198754


# extreme heat day:
finally we choose the 20230819(0.726) and 20160807(0.825)
# normal heat day:
then I need to check the normal heat day percentile (40-60%) during the recent 30 years （1995-2024）

In [24]:
import os
import pandas as pd

input_dir = r'D:\seoul\a_airtem'
csv_files = [
    os.path.join(input_dir, 'AT_1995_2004.csv'),
    os.path.join(input_dir, 'AT_2005_2014.csv'),
    os.path.join(input_dir, 'AT_2015_2024.csv'),
]

# 读取并合并
dfs = [pd.read_csv(f, skiprows=3, encoding='euc_kr') for f in csv_files]
tem = pd.concat(dfs, ignore_index=True)

# 查看合并结果
print(tem.head())
tem.to_csv(r'D:\seoul\a_airtem\result\tem.csv')

           일자  기온(°C)  습도(%rh)  체감온도(°C)
0  1995-05-01    25.6     25.0      23.2
1  1995-05-02    22.0     40.0      21.4
2  1995-05-03    16.8     31.0      15.9
3  1995-05-04    17.2     23.0      15.6
4  1995-05-05    17.8     42.0      17.6


In [25]:
file = r'D:\seoul\a_airtem\result\tem.csv'
data = pd.read_csv(file)
data
# 计算每行 AT(체감온도(°C)) 的百分等级（0~100）
data['percentile'] = data['체감온도(°C)'].rank(pct=True) * 100

# 只保留你关心的几列
out_cols = ['일자', '기온(°C)', '습도(%rh)', '체감온도(°C)', 'percentile']
data_out = data[out_cols]

# 6. 保存为 UTF-8-SIG 编码的 CSV
out_fp = r'D:\seoul\a_airtem\result\temps.csv'
data_out.to_csv(out_fp, index=False, encoding='utf-8-sig')
print(f"✔ 已保存 {len(data_out)} 条 5/6/7/8/9 月记录（含 percentile 列）到：{out_fp}")


✔ 已保存 4590 条 5/6/7/8/9 月记录（含 percentile 列）到：D:\seoul\a_airtem\result\temps.csv


In [26]:
file = r'D:\seoul\a_airtem\result\temps.csv'
data = pd.read_csv(file)
data

,일자,기온(°C),습도(%rh),체감온도(°C),percentile
0,1995-05-01,25.6,25.0,23.2,13.823529
1,1995-05-02,22.0,40.0,21.4,7.287582
2,1995-05-03,16.8,31.0,15.9,0.152505
3,1995-05-04,17.2,23.0,15.6,0.108932
4,1995-05-05,17.8,42.0,17.6,0.784314
...,...,...,...,...,...
4585,2024-09-26,28.0,57.9,28.3,52.875817
4586,2024-09-27,27.7,57.1,27.9,49.226580
4587,2024-09-28,27.9,51.7,27.6,46.448802
4588,2024-09-29,28.9,49.0,28.4,53.954248


In [38]:
file = r'D:\seoul\a_airtem\result\temps.csv'
data = pd.read_csv(file)
data

# 条件筛选：百分位在 40 到 60 之间的行
filtered_data = data[(data['percentile'] >= 40) & (data['percentile'] <= 60)]

# 打印这些行的 '일시' 列
print(filtered_data[['일자', 'percentile']])


              일자  percentile
38    1995-06-08   48.409586
39    1995-06-09   49.226580
40    1995-06-10   53.954248
43    1995-06-13   55.163399
46    1995-06-16   44.542484
...          ...         ...
4585  2024-09-26   52.875817
4586  2024-09-27   49.226580
4587  2024-09-28   46.448802
4588  2024-09-29   53.954248
4589  2024-09-30   47.505447

[926 rows x 2 columns]


In [39]:
# 转换日期格式：去掉 "-"，筛选以 2024 或 2016 开头的日期
sorted_dates = [str(d).replace("-", "") for d in filtered_data['일자']]
filtered_dates = [d for d in sorted_dates if d.startswith('2023') or d.startswith('2016')]

# 用 boolean mask 获取对应的行（原始 '일시' 格式）
filtered_rows = filtered_data[filtered_data['일자'].astype(str).str.replace("-", "").isin(filtered_dates)]

# 打印日期和 percentile 两列
filtered_rows[['일자', 'percentile']]


,일자,percentile
3231,2016-05-19,55.163399
3234,2016-05-22,52.875817
3235,2016-05-23,45.413943
3240,2016-05-28,41.808279
3242,2016-05-30,44.542484
...,...,...
4406,2023-08-31,57.592593
4420,2023-09-14,49.226580
4422,2023-09-16,55.163399
4425,2023-09-19,52.026144


In [40]:
import numpy as np
import pandas as pd

# 将 '일시' 列转换为字符串形式的日期
normal_dates_clean = pd.to_datetime(filtered_data['일자']).dt.strftime('%Y-%m-%d')
filtered_dates = [d for d in normal_dates_clean if d.startswith('2023') or d.startswith('2016')]

set_normal_array = np.array(filtered_dates)

# 将 acquired_clean 转为字符串形式
acquired_str_array = np.array([str(d)[:10] for d in acquired_clean])  # 保留到日，防止有时间部分

# 取交集
common = np.intersect1d(acquired_str_array, set_normal_array)

# 排序并转成列表
common_sorted = sorted(common.tolist())
print(f"共有 {len(common_sorted)} 个相同的日期：{common_sorted}")


共有 11 个相同的日期：['2016-05-19', '2016-05-28', '2016-07-06', '2016-09-08', '2016-09-17', '2016-09-24', '2023-05-31', '2023-06-09', '2023-06-16', '2023-08-28', '2023-09-28']


In [41]:
import os
import rasterio
import geopandas as gpd
from rasterio.mask import mask
from glob import glob

def clip_rasters_with_shapefile(raster_folder, shapefile_path, output_folder):
    # Ensure the output directory exists
    os.makedirs(output_folder, exist_ok=True)

    # Load the shapefile as a GeoDataFrame
    nyc_shape = gpd.read_file(shapefile_path)

    # Loop through each raster in the raster folder
    for raster_path in glob(os.path.join(raster_folder, '*.tif')):
        with rasterio.open(raster_path) as src:
            raster_crs = src.crs

            # Ensure the shapefile is in the same CRS as the raster
            if nyc_shape.crs != raster_crs:
                nyc_shape = nyc_shape.to_crs(raster_crs)

            # Clip the raster using the shapefile
            out_image, out_transform = mask(src, nyc_shape.geometry, crop=True)

            # Update metadata for the output raster
            out_meta = src.meta.copy()
            out_meta.update({
                "driver": "GTiff",
                "height": out_image.shape[1],
                "width": out_image.shape[2],
                "transform": out_transform,
                "crs": raster_crs  # Use the raster's CRS for the output
            })

            # Save the clipped raster to the output folder
            output_raster_path = os.path.join(output_folder, f"clipped_{os.path.basename(raster_path)}")
            if os.path.isdir(output_raster_path):
                continue
            with rasterio.open(output_raster_path, "w", **out_meta) as dest:
                dest.write(out_image)

            print(f"Clipped raster saved to: {output_raster_path}")

In [31]:
# Paths to the folder and shapefile
raster_folder = r'D:\seoul\b_satellite_img\normal\original'
seoul_boundary_file = r'D:\seoul\Final_data\Admin_boundary\Seoul_boundary.shp'
shp_seoul = seoul_boundary_file
output_folder = r'D:\seoul\b_satellite_img\normal\clipped'
os.makedirs(output_folder, exist_ok=True)

clip_rasters_with_shapefile(raster_folder, shp_seoul, output_folder)

Clipped raster saved to: D:\seoul\b_satellite_img\normal\clipped\clipped_LC08_L2SP_115034_20160528_20200906_02_T1_QA_PIXEL.TIF
Clipped raster saved to: D:\seoul\b_satellite_img\normal\clipped\clipped_LC08_L2SP_115034_20160528_20200906_02_T1_ST_B10.TIF
Clipped raster saved to: D:\seoul\b_satellite_img\normal\clipped\clipped_LC08_L2SP_116034_20160519_20200907_02_T1_QA_PIXEL.TIF
Clipped raster saved to: D:\seoul\b_satellite_img\normal\clipped\clipped_LC08_L2SP_116034_20160519_20200907_02_T1_ST_B10.TIF
Clipped raster saved to: D:\seoul\b_satellite_img\normal\clipped\clipped_LC08_L2SP_116034_20160706_20200906_02_T1_QA_PIXEL.TIF
Clipped raster saved to: D:\seoul\b_satellite_img\normal\clipped\clipped_LC08_L2SP_116034_20160706_20200906_02_T1_ST_B10.TIF
Clipped raster saved to: D:\seoul\b_satellite_img\normal\clipped\clipped_LC08_L2SP_116034_20160908_20200906_02_T1_QA_PIXEL.TIF
Clipped raster saved to: D:\seoul\b_satellite_img\normal\clipped\clipped_LC08_L2SP_116034_20160908_20200906_02_T1_ST_

In [42]:
import os
import rasterio
import numpy as np
import pandas as pd
import glob
# Mapping from month names to numbers for date comparison
month_to_num = {
    'Jan': '01', 'Feb': '02', 'Mar': '03', 'Apr': '04', 'May': '05', 'Jun': '06',
    'Jul': '07', 'Aug': '08', 'Sep': '09', 'Oct': '10', 'Nov': '11', 'Dec': '12'
}

def calculate_pixel_proportion(sorted_dates, base_folder, pixel_values=[21824, 21952]):
    for sorted_data in sorted_dates:
        pattern = r'{}\*_*{}_*_QA_PIXEL.TIF'.format(base_folder, sorted_data)
        matched_files = glob.glob(pattern)
        if not matched_files:
            continue
        with rasterio.open(matched_files[0]) as src:
            qa_pixel_data = src.read(1)
        non_zero_pixels = qa_pixel_data[qa_pixel_data > 0]
        total_non_zero_pixels = non_zero_pixels.size
        unique, counts = np.unique(non_zero_pixels, return_counts=True)
        pixel_counts = dict(zip(unique, counts))
        pixel_proportions = {value: count / total_non_zero_pixels for value, count in pixel_counts.items()}

        print(f"📅 {sorted_data}:")
        val_all = 0
        for val in pixel_values:
            prop = pixel_proportions.get(val, 0)
            print(f"  - 像素值 {val} 的占比: {prop:.2%}")
            val_all = val_all + prop
        print(val_all)
        


# Example usage
base_folder = r'D:\seoul\b_satellite_img\normal\clipped'

calculate_pixel_proportion(sorted_dates, base_folder)

📅 20160519:
  - 像素值 21824 的占比: 95.42%
  - 像素值 21952 的占比: 4.50%
0.999201979447627
📅 20160528:
  - 像素值 21824 的占比: 7.65%
  - 像素值 21952 的占比: 0.02%
0.07674223001388311
📅 20160706:
  - 像素值 21824 的占比: 15.72%
  - 像素值 21952 的占比: 0.28%
0.15999494735590677
📅 20160908:
  - 像素值 21824 的占比: 22.72%
  - 像素值 21952 的占比: 0.75%
0.23469234598723465
📅 20160924:
  - 像素值 21824 的占比: 94.84%
  - 像素值 21952 的占比: 4.62%
0.9946189340407035
📅 20230531:
  - 像素值 21824 的占比: 83.65%
  - 像素值 21952 的占比: 2.50%
0.8614624432506334
📅 20230616:
  - 像素值 21824 的占比: 94.80%
  - 像素值 21952 的占比: 4.21%
0.9900418329209484
📅 20230928:
  - 像素值 21824 的占比: 27.93%
  - 像素值 21952 的占比: 1.85%
0.2977850099938328


# 确定了 extreme heat and normal heat days
extreme heat 20160807, 20230819
normal heat 20160924, 20230616

# 下一步是得到 heat resilience map

# 用 QA数据清理B10 -> clean data

In [ ]:
import rasterio
import numpy as np
import os
import glob

extreme_heat=[20160807, 20230819]
normal_heat=[20160924, 20230616]
label_dict = {d: 'extreme' for d in extreme_heat}
label_dict.update({d: 'normal' for d in normal_heat})

for date, label in label_dict.items():
    print(f"{label} heat day:", date)
    data_workspace = rf"D:\seoul\b_satellite_img\{label}\clipped"
    b_10_data = glob.glob(os.path.join(data_workspace, rf'clipped_*_{date}_*_ST_B10.TIF'))
    qa_data = glob.glob(os.path.join(data_workspace, rf'clipped_*_{date}_*_QA_PIXEL.TIF'))
    output_folder = os.path.join(data_workspace, r'clean')
    os.makedirs(output_folder, exist_ok=True)

    if not b_10_data or not qa_data:
        print(f"No file found for date {date}")
        continue

    b_10_data = b_10_data[0]
    qa_data = qa_data[0]

    output_data = os.path.join(output_folder, rf'clipped_{date}_B10_cleaned.tif')
    print("B10 Data Path:", b_10_data)
    print("QA Data Path:", qa_data)
    print("Output Data Path:", output_data)

    # Read Band 10 and QA data
    with rasterio.open(b_10_data) as src:
        band10 = src.read(1).astype(float)
        profile = src.profile

    with rasterio.open(qa_data) as qa_src:
        qa = qa_src.read(1)

    # QA filtering
    valid_labels = [21824, 21952]
    cloud_mask = ~np.isin(qa, valid_labels)
    band10[cloud_mask] = 0

    with rasterio.open(output_data, 'w', **profile) as dst:
        dst.write(band10, 1)



# raster 转 shapefile 

In [55]:
import os
import rasterio
import geopandas as gpd
from rasterio.features import shapes
from shapely.geometry import shape

def raster_to_shapefile(raster_files, output_folder):
    os.makedirs(output_folder, exist_ok=True)

    for raster_file in raster_files:
        with rasterio.open(raster_file) as src:
            nodata_value = src.nodata
            output_shapefile = os.path.join(
                output_folder,
                os.path.basename(raster_file).replace('.tif', '.shp')
            )

            if os.path.exists(output_shapefile):
                print(f"⏩ 已存在，跳过: {output_shapefile}")
                continue

            mask = src.read(1)
            results = shapes(mask, mask=mask != nodata_value, transform=src.transform)

            # ➕ 创建 GeoDataFrame
            shapes_list = []
            for geom, value in results:
                geom = shape(geom)
                if geom.is_valid:
                    shapes_list.append({"geometry": geom, "value": value})

            gdf = gpd.GeoDataFrame(shapes_list, crs=src.crs)

            # ✅ 保存
            gdf.to_file(output_shapefile)
            print(f"✅ 栅格已转换为 Shapefile 并保存到: {output_shapefile}")

# extreme_heat=[20160807, 20230819]
# normal_heat=[20160924, 20230616]
# # 输入的栅格文件路径
# normal_files = [
#     rf'D:\seoul\b_satellite_img\\normal\clipped\clean\clipped_{normal_heat[0]}_B10_cleaned.tif',
#     rf'D:\seoul\b_satellite_img\\normal\clipped\clean\clipped_{normal_heat[1]}_B10_cleaned.tif'
# ]
# extreme_files = [
#     rf'D:\seoul\b_satellite_img\extreme\clipped\clean\clipped_{extreme_heat[0]}_B10_cleaned.tif',
#     rf'D:\seoul\b_satellite_img\extreme\clipped\clean\clipped_{extreme_heat[1]}_B10_cleaned.tif'
#  ]
# 合并所有栅格文件
# all_raster_files = normal_files + extreme_files
all_raster_files_folder = r'D:\seoul\b_satellite_img\heat_resilience'
# 获取文件夹中所有.tif栅格文件的完整路径
all_raster_files = [
    os.path.join(all_raster_files_folder, file)
    for file in os.listdir(all_raster_files_folder)
    if file.endswith('.tif')
]
# 输出文件夹路径
output_shapefile_folder = r'D:\seoul\b_satellite_img\shapefiles'
os.makedirs(output_shapefile_folder, exist_ok=True)

# 转换栅格文件为 Shapefile
raster_to_shapefile(all_raster_files, output_shapefile_folder)

⏩ 已存在，跳过: D:\seoul\b_satellite_img\shapefiles\2016_heat_resilience.shp
⏩ 已存在，跳过: D:\seoul\b_satellite_img\shapefiles\2023_heat_resilience.shp


# 得到fid的shape文件

In [56]:
import geopandas as gpd
import pandas as pd
import os
import glob

def fast_clip_and_assign_fid(input_shp, clip_shp, output_shp):
    print(f"📂 加载输入数据：{input_shp}")
    input_gdf = gpd.read_file(input_shp)
    clip_gdf = gpd.read_file(clip_shp)

    # 确保 CRS 一致
    if input_gdf.crs != clip_gdf.crs:
        input_gdf = input_gdf.to_crs(clip_gdf.crs)

    # 在裁剪前先加上 city_id 字段
    clip_gdf = clip_gdf.reset_index().rename(columns={'index': 'city_id'})

    print("✂️ 正在执行 overlay 精准裁剪...")
    try:
        clipped = gpd.overlay(input_gdf, clip_gdf[['city_id', 'geometry']], how='intersection', keep_geom_type=False)
    except Exception as e:
        print(f"❌ Overlay 错误: {e}")
        return

    # 保存结果
    os.makedirs(os.path.dirname(output_shp), exist_ok=True)
    clipped.to_file(output_shp, driver='ESRI Shapefile')
    print(f"✅ 已保存: {output_shp}")
# extreme_heat=[20160807, 20230819]
# normal_heat=[20160924, 20230616]

# # 输入的栅格转 Shapefile 路径（假设已经将其转化为 Shapefile）
# shapefile_files = [
#     rf'D:\seoul\b_satellite_img\shapefiles\clipped_{normal_heat[0]}_B10_cleaned.shp',
#     rf'D:\seoul\b_satellite_img\shapefiles\clipped_{normal_heat[1]}_B10_cleaned.shp',
#     rf'D:\seoul\b_satellite_img\shapefiles\clipped_{extreme_heat[0]}_B10_cleaned.shp',
#     rf'D:\seoul\b_satellite_img\shapefiles\clipped_{extreme_heat[1]}_B10_cleaned.shp'

# ]
shapefile_files_folder = r'D:\seoul\b_satellite_img\shapefiles'
# 获取文件夹中所有.tif栅格文件的完整路径
shapefile_files = [
    os.path.join(shapefile_files_folder, file)
    for file in os.listdir(shapefile_files_folder)
    if file.endswith('.shp')
]
grid_folder = r'D:\seoul\grids'
# 输出文件夹路径
output_folder = r'D:\seoul\b_satellite_img\shapefiles_clipped_with_fid'
# 遍历所有 grid shapefile
for grid_file in os.listdir(grid_folder):
    if grid_file.endswith('.shp'):
        grid_path = os.path.join(grid_folder, grid_file)

        for input_shp in shapefile_files:
            input_name = os.path.splitext(os.path.basename(input_shp))[0]
            grid_name = os.path.splitext(os.path.basename(grid_file))[0]
            output_name = f'{input_name}_clipped_by_{grid_name}_with_fid.shp'
            output_shp = os.path.join(output_folder, output_name)
            if os.path.exists(output_shp):
                print(f"⏩ 已存在，跳过: {output_shp}")
                continue
            fast_clip_and_assign_fid(input_shp, grid_path, output_shp)

⏩ 已存在，跳过: D:\seoul\b_satellite_img\shapefiles_clipped_with_fid\2016_heat_resilience_clipped_by_grid_1080m_with_fid.shp
⏩ 已存在，跳过: D:\seoul\b_satellite_img\shapefiles_clipped_with_fid\2023_heat_resilience_clipped_by_grid_1080m_with_fid.shp
⏩ 已存在，跳过: D:\seoul\b_satellite_img\shapefiles_clipped_with_fid\2016_heat_resilience_clipped_by_grid_120m_with_fid.shp
⏩ 已存在，跳过: D:\seoul\b_satellite_img\shapefiles_clipped_with_fid\2023_heat_resilience_clipped_by_grid_120m_with_fid.shp
⏩ 已存在，跳过: D:\seoul\b_satellite_img\shapefiles_clipped_with_fid\2016_heat_resilience_clipped_by_grid_240m_with_fid.shp
⏩ 已存在，跳过: D:\seoul\b_satellite_img\shapefiles_clipped_with_fid\2023_heat_resilience_clipped_by_grid_240m_with_fid.shp
⏩ 已存在，跳过: D:\seoul\b_satellite_img\shapefiles_clipped_with_fid\2016_heat_resilience_clipped_by_grid_360m_with_fid.shp
⏩ 已存在，跳过: D:\seoul\b_satellite_img\shapefiles_clipped_with_fid\2023_heat_resilience_clipped_by_grid_360m_with_fid.shp
⏩ 已存在，跳过: D:\seoul\b_satellite_img\shapefiles_clipped_

# 先将数据洗成摄氏度 bt -> lst

In [57]:
import rasterio
import numpy as np
import os


def calculate_lst(input_files, output_dir, multiplier=0.00341802, add_constant=149.0, nodata_value=-9999):
    """
    Calculate Land Surface Temperature (LST) from a list of raster files and save the output.

    Args:
        input_files (list): List of paths to input raster files.
        output_dir (str): Directory to save the output raster files.
        multiplier (float): Multiplication factor for converting to temperature.
        add_constant (float): Additive constant for converting to temperature.
        nodata_value (float): Value to use for no-data in the output files.
    """
    # Ensure the output directory exists
    os.makedirs(output_dir, exist_ok=True)

    for input_file in input_files:
        with rasterio.open(input_file) as src:
            # Read the first band and convert to float
            band_data = src.read(1).astype(float)
            profile = src.profile

        # Replace values <= 0 with NaN
        band_data[band_data <= 0] = np.nan

        # Calculate brightness temperature (BT)
        lst_kelvin = (band_data * multiplier) + add_constant

        # Convert to Celsius
        lst_celsius = lst_kelvin - 273.15

        # Replace NaN values with the nodata value
        lst_celsius[np.isnan(lst_celsius)] = nodata_value

        # Update profile for the output file
        profile.update({
            'dtype': 'float32',
            'count': 1,
            'nodata': nodata_value
        })

        # Generate a short output file name
        base_name = os.path.basename(input_file).split('.')[0]
        output_file = os.path.join(output_dir, f"{base_name}_LST.tif")

        # Save the LST raster
        with rasterio.open(output_file, 'w', **profile) as dst:
            dst.write(lst_celsius, 1)

        print(f"LST saved to: {output_file}")

extreme_heat=[20160807, 20230819]
normal_heat=[20160924, 20230616]
# 输入的栅格文件路径
normal_files = [
    rf'D:\seoul\b_satellite_img\\normal\clipped\clean\clipped_{normal_heat[0]}_B10_cleaned.tif',
    rf'D:\seoul\b_satellite_img\\normal\clipped\clean\clipped_{normal_heat[1]}_B10_cleaned.tif'
]
extreme_files = [
    rf'D:\seoul\b_satellite_img\extreme\clipped\clean\clipped_{extreme_heat[0]}_B10_cleaned.tif',
    rf'D:\seoul\b_satellite_img\extreme\clipped\clean\clipped_{extreme_heat[1]}_B10_cleaned.tif'
]

# 输出目录
output_dir_normal = r'D:\seoul\b_satellite_img\\normal_heat_day_lst'
output_dir_extreme = r'D:\seoul\b_satellite_img\\extreme_heat_day_lst'

# 计算并保存 LST
calculate_lst(normal_files, output_dir_normal)
calculate_lst(extreme_files, output_dir_extreme)

LST saved to: D:\seoul\b_satellite_img\\normal_heat_day_lst\clipped_20160924_B10_cleaned_LST.tif
LST saved to: D:\seoul\b_satellite_img\\normal_heat_day_lst\clipped_20230616_B10_cleaned_LST.tif
LST saved to: D:\seoul\b_satellite_img\\extreme_heat_day_lst\clipped_20160807_B10_cleaned_LST.tif
LST saved to: D:\seoul\b_satellite_img\\extreme_heat_day_lst\clipped_20230819_B10_cleaned_LST.tif


# 相减，然后得到HR map

In [60]:
import rasterio
import numpy as np
import os

def calculate_heat_resilience(normal_files, extreme_files, output_dir, nodata_value=-9999):
    os.makedirs(output_dir, exist_ok=True)

    if len(normal_files) != len(extreme_files):
        raise ValueError("Number of normal and extreme files must be the same.")

    for normal_file, extreme_file in zip(normal_files, extreme_files):
        with rasterio.open(normal_file) as normal_src, rasterio.open(extreme_file) as extreme_src:
            normal_data = normal_src.read(1).astype(float)
            extreme_data = extreme_src.read(1).astype(float)

            if normal_data.shape != extreme_data.shape:
                raise ValueError(f"Dimension mismatch between {normal_file} and {extreme_file}.")

            # 强制处理 nodata
            if normal_src.nodata is not None:
                normal_data[normal_data == normal_src.nodata] = np.nan
            else:
                normal_data[normal_data == nodata_value] = np.nan

            if extreme_src.nodata is not None:
                extreme_data[extreme_data == extreme_src.nodata] = np.nan
            else:
                extreme_data[extreme_data == nodata_value] = np.nan

            # 差值
            heat_resilience = normal_data - extreme_data

            # 打印调试
            print(f"\n📁 Processing: {os.path.basename(normal_file)} vs {os.path.basename(extreme_file)}")
            print("🔹 First 10 values from normal_data (non-NaN):", normal_data[~np.isnan(normal_data)].flatten()[:10])
            print("🔹 First 10 values from extreme_data (non-NaN):", extreme_data[~np.isnan(extreme_data)].flatten()[:10])
            print("🔸 First 10 values from heat_resilience (non-NaN):", heat_resilience[~np.isnan(heat_resilience)].flatten()[:10])

            # 替换 NaN 为 nodata
            heat_resilience[np.isnan(heat_resilience)] = nodata_value

            profile = normal_src.profile
            profile.update({
                'dtype': 'float32',
                'nodata': nodata_value
            })

            year = os.path.basename(normal_file).split('_')[1][:4]
            output_file = os.path.join(output_dir, f"{year}_heat_resilience.tif")

            with rasterio.open(output_file, 'w', **profile) as dst:
                dst.write(heat_resilience.astype('float32'), 1)

            print(f"✅ Heat resilience raster saved to: {output_file}")


extreme_heat=[20160807, 20230819]
normal_heat=[20160924, 20230616]
# 输入的栅格文件路径
normal_files_lst = [
    rf'D:\seoul\b_satellite_img\normal_heat_day_lst\clipped_{normal_heat[0]}_B10_cleaned_lst.tif',
    rf'D:\seoul\b_satellite_img\normal_heat_day_lst\clipped_{normal_heat[1]}_B10_cleaned_lst.tif'
]
extreme_files_lst = [
    rf'D:\seoul\b_satellite_img\extreme_heat_day_lst\clipped_{extreme_heat[0]}_B10_cleaned_lst.tif',
    rf'D:\seoul\b_satellite_img\extreme_heat_day_lst\clipped_{extreme_heat[1]}_B10_cleaned_lst.tif'
]

# 输出目录
output_dir = r'D:\seoul\b_satellite_img\heat_resilience'

# 计算并保存 heat resilience
calculate_heat_resilience(normal_files_lst, extreme_files_lst, output_dir, nodata_value=-9999)


📁 Processing: clipped_20160924_B10_cleaned_lst.tif vs clipped_20160807_B10_cleaned_lst.tif
🔹 First 10 values from normal_data (non-NaN): [23.6862011  24.06218338 23.54606247 23.81266785 23.80583191 24.33562469
 24.08952713 24.38005829 24.65008354 24.62957382]
🔹 First 10 values from extreme_data (non-NaN): [32.88751221 32.70977402 32.58330536 32.68584824 35.6390152  36.1790657
 36.73620224 34.06672668 34.63412094 35.2767067 ]
🔸 First 10 values from heat_resilience (non-NaN): [-6.46005821 -6.33359146 -6.21737671 -6.35751724 -7.13340569 -7.32481766
 -7.54698944 -7.05821037 -7.19835281 -7.35899734]
✅ Heat resilience raster saved to: D:\seoul\b_satellite_img\heat_resilience\2016_heat_resilience.tif

📁 Processing: clipped_20230616_B10_cleaned_lst.tif vs clipped_20230819_B10_cleaned_lst.tif
🔹 First 10 values from normal_data (non-NaN): [26.51290321 27.26828575 26.66671371 27.00509834 26.90255737 27.70921135
 27.27512169 27.61008835 27.86302185 27.91087341]
🔹 First 10 values from extreme_data

# 计算面积和占比

In [ ]:
import geopandas as gpd
import os
from concurrent.futures import ThreadPoolExecutor

def add_lst_ratio_within_only(clipped_gdf, city_path, output_path):
    city_gdf = gpd.read_file(city_path)
    target_crs = city_gdf.crs
    clipped = clipped_gdf.to_crs(target_crs)
    city_gdf = city_gdf.to_crs(target_crs)

    city_gdf = city_gdf.reset_index(drop=True)
    city_gdf['city_id'] = city_gdf.index
    city_gdf['lst'] = 0.0  # 初始化 lst 值

    for idx, row in city_gdf.iterrows():
        grid_geom = row.geometry
        inside = clipped[clipped.within(grid_geom)]
        if not inside.empty:
            area_sum = inside.geometry.area.sum()
            city_gdf.at[idx, 'lst'] = area_sum

    # 确保有 Shape_Area 字段，否则计算它
    if 'Shape_Area' not in city_gdf.columns:
        city_gdf['Shape_Area'] = city_gdf.geometry.area

    city_gdf['lst_ratio'] = city_gdf['lst'] / city_gdf['Shape_Area']

    # 输出为 ITRF_2000_UTM_K 坐标系
    itrf_proj = (
        '+proj=tmerc +lat_0=38 +lon_0=127.5 '
        '+k=0.9996 +x_0=1000000 +y_0=2000000 '
        '+ellps=GRS80 +units=m +no_defs'
    )
    city_gdf_final = city_gdf.to_crs(itrf_proj)
    city_gdf_final.to_file(output_path, driver='ESRI Shapefile')

    print(f"✅ 完全包含统计完成并保存: {output_path}")


# === 设置路径 ===
clipped_folder = r'D:\seoul\b_satellite_img\shapefiles_clipped_with_fid'
grid_folder = r'D:\seoul\grids'
output_folder = r'D:\seoul\grids\city_ratio_outputs'
os.makedirs(output_folder, exist_ok=True)

# 所有文件
clipped_files = [f for f in os.listdir(clipped_folder) if f.endswith('.shp')]
grid_files = [f for f in os.listdir(grid_folder) if f.endswith('.shp')]

def process_pair(grid_file):
    try:
        grid_name = os.path.splitext(grid_file)[0]
        grid_path = os.path.join(grid_folder, grid_file)

        # 匹配当前 grid 的所有年份版本 clipped 文件（可能有 2016 和 2023）
        matched = [f for f in clipped_files if grid_name in f]
        if not matched:
            print(f"⚠️ 没有找到匹配 clipped 文件: {grid_name}")
            return

        for match in matched:
            clipped_path = os.path.join(clipped_folder, match)
            clipped_gdf = gpd.read_file(clipped_path)
            clipped_gdf = clipped_gdf.reset_index(drop=True)
            clipped_gdf['city_id'] = clipped_gdf['FID'] if 'FID' in clipped_gdf.columns else clipped_gdf.index

            # 判断年份
            year = "2023" if "2023" in match else "2016" if "2016" in match else "unknown"
            if year == "unknown":
                print(f"⚠️ 未识别年份，跳过: {match}")
                continue

            output_path = os.path.join(output_folder, f'city{year}_lst_ratio_{grid_name}.shp')
            if os.path.exists(output_path):
                print(f"⏩ 已存在，跳过: {output_path}")
                continue

            add_lst_ratio_within_only(clipped_gdf, grid_path, output_path)

    except Exception as e:
        print(f"❌ 处理 {grid_file} 失败: {e}")

# 🚀 并行执行
print("🚀 开始批量处理 grid ↔ clipped 匹配...")
with ThreadPoolExecutor(max_workers=4) as executor:
    executor.map(process_pair, grid_files)


🚀 开始批量处理 grid ↔ clipped 匹配...
⏩ 已存在，跳过: D:\seoul\grids\city_ratio_outputs\city2016_lst_ratio_grid_1080m.shp
✅ 完全包含统计完成并保存: D:\seoul\grids\city_ratio_outputs\city2016_lst_ratio_grid_360m.shp
✅ 完全包含统计完成并保存: D:\seoul\grids\city_ratio_outputs\city2023_lst_ratio_grid_1080m.shp
⏩ 已存在，跳过: D:\seoul\grids\city_ratio_outputs\city2016_lst_ratio_grid_480m.shp
✅ 完全包含统计完成并保存: D:\seoul\grids\city_ratio_outputs\city2023_lst_ratio_grid_480m.shp
✅ 完全包含统计完成并保存: D:\seoul\grids\city_ratio_outputs\city2023_lst_ratio_grid_360m.shp
⏩ 已存在，跳过: D:\seoul\grids\city_ratio_outputs\city2016_lst_ratio_grid_600m.shp
✅ 完全包含统计完成并保存: D:\seoul\grids\city_ratio_outputs\city2016_lst_ratio_grid_720m.shp
✅ 完全包含统计完成并保存: D:\seoul\grids\city_ratio_outputs\city2023_lst_ratio_grid_720m.shp
✅ 完全包含统计完成并保存: D:\seoul\grids\city_ratio_outputs\city2023_lst_ratio_grid_600m.shp
✅ 完全包含统计完成并保存: D:\seoul\grids\city_ratio_outputs\city2016_lst_ratio_grid_240m.shp
✅ 完全包含统计完成并保存: D:\seoul\grids\city_ratio_outputs\city2016_lst_ratio_grid_840m.shp

# ** 建筑清理

In [ ]:
import geopandas as gpd

file = r'D:\seoul\c_data\a_building\final\18-3.Fin_building height\18-3.Fin_building height.shp'
data = gpd.read_file(file)
data

,명칭,구분,종류,용도,주기,층수,측량방,UFID,height,area,geometry
0,None,None,연립주택,주택,힐스테이트V,5,사진측량,1000037705092B00110000000000008452,15.0,134.759745,"POLYGON ((327581.616 4153675.479, 327584.987 4..."
1,None,None,연립주택,주택,욱현빌라,3,사진측량,1000037705074B00110000000000002733,9.0,90.559652,"POLYGON ((331526.206 4158037.674, 331518.521 4..."
2,None,None,주택외건물,근린생활시설,신당로얄빌라경비실,1,사진측량,1000037705071B00110000000000018407,3.0,9.181550,"POLYGON ((324545.839 4158405.230, 324543.741 4..."
3,None,None,연립주택,주택,금강빌라트,4,사진측량,1000037705061B00110000000000016459,12.0,282.726302,"POLYGON ((325279.430 4162587.526, 325275.990 4..."
4,None,None,주택외건물,업무시설,한국동물병원협회동물사업본부,6,사진측량,1000037705092B00110000000000007369,18.0,133.262507,"POLYGON ((326916.174 4153310.619, 326913.932 4..."
...,...,...,...,...,...,...,...,...,...,...,...
641420,None,None,주택외건물,공장,서울탁주도봉연합제조장,1,사진측량,1000037705042B00110000000000006875,3.0,111.856802,"POLYGON ((326777.501 4167416.848, 326788.435 4..."
641421,None,None,주택외건물,공장,서울탁주도봉연합제조장,1,사진측량,1000037705042B00110000000000006875,3.0,170.909467,"POLYGON ((326782.546 4167464.502, 326789.775 4..."
641422,None,None,주택외건물,공장,서울탁주도봉연합제조장,1,사진측량,1000037705042B00110000000000006875,3.0,63.583773,"POLYGON ((326782.549 4167474.500, 326772.085 4..."
641423,퍼스트AVENUE타워,None,주택외건물,근린생활시설,"강남퍼스트약국,중앙산부인과",15,사진측량,1000037705091B00110000000000005195,45.0,390.631864,"POLYGON ((325057.053 4153981.922, 325064.931 4..."


In [ ]:
print(data["종류"].unique())
print(data["종류"].value_counts())

['연립주택' '주택외건물' '아파트' '공사중건물' '온실' '일반주택']
종류
일반주택     274554
주택외건물    240674
연립주택      98288
아파트       24719
공사중건물      1854
온실         1336
Name: count, dtype: int64


In [ ]:
import geopandas as gpd

file_2 = r'D:\seoul\bldg_merge\bldg_merge.shp'
data_2 = gpd.read_file(file_2)
print(data_2["종류"].unique())
print(data_2["종류"].value_counts())

['무벽건물' '주택외건물' '일반주택' '연립주택' '공사중건물' '아파트' '가건물' '온실']
종류
일반주택     296666
주택외건물    274915
무벽건물     174490
연립주택     108659
아파트       28915
가건물       21775
온실         4816
공사중건물      2251
Name: count, dtype: int64


In [ ]:
import geopandas as gpd

# 读取数据
file = r'D:\seoul\c_data\a_building\final\18-3.Fin_building height\18-3.Fin_building height.shp'
data = gpd.read_file(file)

# 过滤掉 area 小于 5 的记录
filtered_data = data[data['area'] >= 5]

# 保存结果
output_file = r'D:\seoul\c_data\a_building\final\Fin_building.shp'
filtered_data.to_file(output_file, encoding='utf-8')  # encoding 很重要

In [ ]:
import geopandas as gpd

output_file = r'D:\seoul\c_data\a_building\final\Fin_building.shp'
data_3 = gpd.read_file(output_file)
print(data_3["종류"].unique())
print(data_3["종류"].value_counts())

['연립주택' '주택외건물' '아파트' '공사중건물' '온실' '일반주택']
종류
일반주택     271724
주택외건물    238633
연립주택      98280
아파트       24716
공사중건물      1848
온실         1306
Name: count, dtype: int64


# *** 得到HR lst temperature 在extreme heat & normal heat & heat reslience 的数据

In [50]:
import os
import rasterio
import numpy as np
import geopandas as gpd
from rasterio.mask import mask
from shapely.geometry import mapping

def calculate_polygon_means(raster_path, vector_path):
    """
    根据矢量多边形计算栅格的平均值。
    :param raster_path: 栅格文件路径。
    :param vector_path: 矢量文件路径。
    :return: 列名（基于日期）和对应的多边形平均值列表。
    """
    try:
        # 提取日期部分作为列名
        filename = os.path.basename(raster_path)
        date_str = filename.split('_')[2]  # 提取类似 '20170730' 的部分

        # 打开矢量文件
        vector = gpd.read_file(vector_path)

        # 打开栅格文件
        with rasterio.open(raster_path) as raster:
            raster_crs = raster.crs

            # 矢量文件需要投影到栅格的 CRS
            if vector.crs != raster_crs:
                vector = vector.to_crs(raster_crs)
                print(f"Reprojected vector CRS to match raster CRS.")

            # 用于存储每个多边形的平均值
            mean_values = []

            # 遍历每个多边形
            for idx, row in vector.iterrows():
                geom = row['geometry']
                # 使用多边形裁剪栅格
                out_image, out_transform = mask(raster, [mapping(geom)], crop=True, all_touched=True)


                # 如果裁剪区域无数据，赋值为 NaN
                if out_image.size == 0:
                    mean_values.append(np.nan)
                else:
                    data = out_image[0]  # 获取第一波段
                    # # 计算平均值，忽略无效值（如 0 或 NaN）
                    # mean_val = np.nanmean(data[data != raster.nodata])
                    # 获取单个栅格文件的平均值
                    valid_data = data[data != raster.nodata]
                    if valid_data.size > 0:
                        mean_val = np.nanmean(valid_data)
                    else:
                        mean_val = -9999  #或其他你想设定的默认值
                    mean_values.append(mean_val)

            return date_str, mean_values

    except Exception as e:
        print(f"Error processing file {raster_path}: {e}")
        return None, []

# 主函数：处理多个栅格文件并保存到单个文件
def process_rasters_to_single_file(raster_files, vector_path, output_path, group_name, year):
    """
    处理多个栅格文件，将结果添加为矢量文件的新列。
    :param raster_files: 栅格文件路径列表。
    :param vector_path: 矢量文件路径。
    :param output_path: 输出矢量文件路径。
    :param group_name: 当前组的名称（如 normal、extreme 等）。
    :param year: 对应的年份列表。
    """
    try:
        # 打开矢量文件
        vector = gpd.read_file(vector_path)

        # 遍历栅格文件并计算每个多边形的平均值
        for idx, raster_file in enumerate(raster_files):
            date_str, mean_values = calculate_polygon_means(raster_file, vector_path)
            if date_str and mean_values:
                # 添加新列到矢量文件，列名为 group_name_年份
                column_name = f"{group_name}_{year[idx]}"
                vector[column_name] = mean_values
                print(f"Added column: {column_name}")

        # 保存更新后的矢量文件
        vector.to_file(output_path)
        print(f"Saved updated vector file to: {output_path}")

    except Exception as e:
        print(f"Error processing rasters for group {group_name}: {e}")



# 修改日期
# year = '2016'
# # 输入的栅格文件路径
# normal_files = [
#     r'D:\seoul\b_satellite_img\normal_heat_day_lst\clipped_20160924_B10_cleaned_LST.tif'
# ]
# extreme_files = [
#     r'D:\seoul\b_satellite_img\extreme_heat_day_lst\clipped_20160807_B10_cleaned_LST.tif'
# ]
# hr_files = [
#     rf'D:\seoul\b_satellite_img\heat_resilience\{year}_heat_resilience.tif'
# ]
year = '2023'
# 输入的栅格文件路径
normal_files = [
    r'D:\seoul\b_satellite_img\normal_heat_day_lst\clipped_20230616_B10_cleaned_LST.tif'
]
extreme_files = [
    r'D:\seoul\b_satellite_img\extreme_heat_day_lst\clipped_20230819_B10_cleaned_LST.tif'
]
hr_files = [
    rf'D:\seoul\b_satellite_img\heat_resilience\{year}_heat_resilience.tif'
]
# 主脚本
input_folder = r'D:\seoul\grids\city_ratio_outputs'
output_folder = r'D:\seoul\grids\lst_map'
os.makedirs(output_folder, exist_ok=True)
# input_folder = r'D:\seoul\grids\lst_map'
# output_folder = r'D:\seoul\grids\lst_map\test'
# 文件列表与名称的映射
file_groups = [normal_files, extreme_files, hr_files]
group_names = ['nor', 'ext', 'hr']
years =[[year],[year],[year]]

for filename in os.listdir(input_folder):
    if filename.endswith('.shp') and filename.startswith(f'city{year}'):
    # if filename.endswith('xy.shp') and filename.startswith(f'city{year}'):
        input_path = os.path.join(input_folder, filename)
        output_path = os.path.join(output_folder, filename)
        # if os.path.exists(output_path):
        #     print(f"⏩ 已存在，跳过: {output_path}")
        #     continue
        # 初始化矢量文件
        final_vector = gpd.read_file(input_path)

        # 循环处理文件
        for i in range(len(file_groups)):
            group_files = file_groups[i]  # 当前组的文件列表
            group_name = group_names[i]  # 当前组的名称
            print(f"Processing {group_name} files...")  
            for idx, raster_file in enumerate(group_files):
                date_str, mean_values = calculate_polygon_means(raster_file, input_path)
                if date_str and mean_values:
                    column_name = f"{group_name}_{years[i][idx]}"
                    final_vector[column_name] = mean_values
                    print(f"Added column: {column_name}")

        # 保存最终结果
        final_vector.to_file(output_path)
        print(f"Saved final vector file to: {output_path}")

Processing nor files...
Reprojected vector CRS to match raster CRS.
Added column: nor_2023
Processing ext files...
Reprojected vector CRS to match raster CRS.
Added column: ext_2023
Processing hr files...
Reprojected vector CRS to match raster CRS.
Added column: hr_2023
Saved final vector file to: D:\seoul\grids\lst_map\test\city2023_lst_ratio_grid_1080m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp
Processing nor files...
Reprojected vector CRS to match raster CRS.
Added column: nor_2023
Processing ext files...
Reprojected vector CRS to match raster CRS.
Added column: ext_2023
Processing hr files...
Reprojected vector CRS to match raster CRS.
Added column: hr_2023
Saved final vector file to: D:\seoul\grids\lst_map\test\city2023_lst_ratio_grid_120m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp
Processing nor files...
Reprojected vector CRS to match raster CRS.
Added column: nor_2023
Processing ext files...
Reprojected vector CRS to match raster CRS.
Added column: ext_2023
Pro

# 【图】 给初始网格出图

In [61]:
import geopandas as gpd
import matplotlib.pyplot as plt
import os

# 首尔行政边界
shp_seoul = r'D:\seoul\Final_data\Admin_boundary\Seoul_boundary.shp'
gdf_seoul = gpd.read_file(shp_seoul)

# 输出文件夹
output_folder = r'D:\seoul\grids\lst_map\figures\grids_gray'
os.makedirs(output_folder, exist_ok=True)
grid_sizes = list(range(120, 1080 + 1, 120))
for i in grid_sizes:
    shp_path = rf'D:\seoul\grids\grid_{i}m.shp'
    gdf = gpd.read_file(shp_path)

    # 如果两者 CRS 不一致，投影到相同坐标系
    if gdf.crs != gdf_seoul.crs:
        gdf_seoul = gdf_seoul.to_crs(gdf.crs)

    fig, ax = plt.subplots(figsize=(8, 8))
    # 再叠加边界线
    gdf_seoul.plot(ax=ax, facecolor='none', edgecolor='black', linewidth=0.8)
    # 先画格网
    gdf.plot(ax=ax, facecolor='lightgray', edgecolor='gray', linewidth=0.3)

    ax.set_title(f'Grid Size: {i}m', fontsize=14)
    ax.axis('off')

    output_path = os.path.join(output_folder, f'grid_{i}m_gray.png')
    plt.savefig(output_path, dpi=300, bbox_inches='tight')
    plt.close()
    print(f"✅ Saved: {output_path}")

✅ Saved: D:\seoul\grids\lst_map\figures\grids_gray\grid_120m_gray.png
✅ Saved: D:\seoul\grids\lst_map\figures\grids_gray\grid_240m_gray.png
✅ Saved: D:\seoul\grids\lst_map\figures\grids_gray\grid_360m_gray.png
✅ Saved: D:\seoul\grids\lst_map\figures\grids_gray\grid_480m_gray.png
✅ Saved: D:\seoul\grids\lst_map\figures\grids_gray\grid_600m_gray.png
✅ Saved: D:\seoul\grids\lst_map\figures\grids_gray\grid_720m_gray.png
✅ Saved: D:\seoul\grids\lst_map\figures\grids_gray\grid_840m_gray.png
✅ Saved: D:\seoul\grids\lst_map\figures\grids_gray\grid_960m_gray.png
✅ Saved: D:\seoul\grids\lst_map\figures\grids_gray\grid_1080m_gray.png


# 02 building - BRC & building average &BHV

In [1]:
import geopandas as gpd
import os
import numpy as np
from concurrent.futures import ThreadPoolExecutor

# === 文件路径 ===
building_file = r'D:\seoul\c_data\a_building\final\Fin_building.shp'
input_folder = r'D:\seoul\grids\lst_map'
output_folder = input_folder

# ✅ 读取一次 building 数据
print("🔄 正在读取建筑数据...")
building_gdf = gpd.read_file(building_file)

def add_building_stats(building_gdf, grid_path, output_path):
    # 使用 grid 的 CRS 作为标准
    grid_gdf = gpd.read_file(grid_path)
    target_crs = grid_gdf.crs
    building_proj = building_gdf.to_crs(target_crs)
    grid_gdf = grid_gdf.to_crs(target_crs)

    # 初始化字段
    grid_gdf = grid_gdf.reset_index(drop=True)
    grid_gdf['grid_id'] = grid_gdf.index
    grid_gdf['BCR(%)'] = 0.0
    grid_gdf['BHA'] = 0.0
    grid_gdf['BHV'] = 0.0

    # 逐个格子处理
    for idx, row in grid_gdf.iterrows():
        grid_geom = row.geometry
        grid_area = grid_geom.area

        # 找到完全在格子内部的建筑
        inside = building_proj[building_proj.within(grid_geom)]

        if not inside.empty:
            total_bldg_area = inside['area'].sum()
            height_std = inside['height'].std()
            if np.isnan(height_std):
                height_std = 0
            height_ave = inside['height'].mean()

            # 计算指标
            grid_gdf.at[idx, 'BCR(%)'] = (total_bldg_area / grid_area) * 100
            grid_gdf.at[idx, 'BHV'] = height_std
            grid_gdf.at[idx, 'BHA'] = height_ave

    # ✅ 修改为 Korea 2000 / Central Belt 1 投影，EPSG:5181
    grid_out = grid_gdf.to_crs("EPSG:5181")

    grid_out.to_file(output_path, driver='ESRI Shapefile')

    print(f"✅ 处理完成并保存: {output_path}")

# === 遍历并处理所有 lst_map 文件 ===
def process_grid(grid_filename):
    if not grid_filename.endswith('m.shp'):
        return
    input_path = os.path.join(input_folder, grid_filename)
    print(input_path)
    output_path = os.path.join(output_folder, f"{os.path.splitext(grid_filename)[0]}_bcr_bhv.shp")
    print(output_path)
    if os.path.exists(output_path):
        print(f"⏩ 已存在，跳过: {output_path}")
        return  
    add_building_stats(building_gdf, input_path, output_path)

# 并行处理
print("🚀 开始批量计算 BCR 和 BHV...")
with ThreadPoolExecutor(max_workers=4) as executor:
    executor.map(process_grid, os.listdir(input_folder))

🔄 正在读取建筑数据...
🚀 开始批量计算 BCR 和 BHV...
D:\seoul\grids\lst_map\city2016_lst_ratio_grid_1080m.shp
D:\seoul\grids\lst_map\city2016_lst_ratio_grid_1080m_bcr_bhv.shp
⏩ 已存在，跳过: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_1080m_bcr_bhv.shp
D:\seoul\grids\lst_map\city2016_lst_ratio_grid_120m.shp
D:\seoul\grids\lst_map\city2016_lst_ratio_grid_120m_bcr_bhv.shp
D:\seoul\grids\lst_map\city2016_lst_ratio_grid_240m.shp
D:\seoul\grids\lst_map\city2016_lst_ratio_grid_240m_bcr_bhv.shp
D:\seoul\grids\lst_map\city2016_lst_ratio_grid_360m.shp
D:\seoul\grids\lst_map\city2016_lst_ratio_grid_360m_bcr_bhv.shp
⏩ 已存在，跳过: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_240m_bcr_bhv.shp
D:\seoul\grids\lst_map\city2016_lst_ratio_grid_480m.shp
D:\seoul\grids\lst_map\city2016_lst_ratio_grid_480m_bcr_bhv.shp
⏩ 已存在，跳过: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_360m_bcr_bhv.shp
D:\seoul\grids\lst_map\city2016_lst_ratio_grid_600m.shp
D:\seoul\grids\lst_map\city2016_lst_ratio_grid_600m_bcr_bhv.shp
⏩ 已存在，跳过: D:\seoul\

# 02 ndvi
20230819 vs 20230616
20160807 vs 20160924 

In [7]:
import rasterio
import numpy as np
import os

# 输入文件夹和输出文件夹
band_folder = r'D:\seoul\c_data\b_ndvi\a_original'
output_folder = r'D:\seoul\c_data\b_ndvi\b_process'
os.makedirs(output_folder, exist_ok=True)

# 获取所有 tif 文件名
all_files = os.listdir(band_folder)

# 筛选出 B4 和 B5 文件
band4_files = [f for f in all_files if f.endswith('SR_B4.TIF')]
band5_files = [f for f in all_files if f.endswith('SR_B5.TIF')]

for i in range(0,len(band4_files)):
    band4_file = band4_files[i]
    band5_file = band5_files[i]
    band4_path = os.path.join(band_folder, band4_file)
    band5_path = os.path.join(band_folder, band5_file)

    # 提取日期
    date_str = band4_file.split('_')[3]
    output_path = os.path.join(output_folder, f'ndvi_{date_str}.tif')

    # 打开波段文件并计算 NDVI
    with rasterio.open(band4_path) as red_src, rasterio.open(band5_path) as nir_src:
        red = red_src.read(1).astype(float)
        nir = nir_src.read(1).astype(float)
        profile = red_src.profile

        ndvi = np.where(
            (nir + red) == 0,
            0,
            (nir - red) / (nir + red)
        )

        profile.update(dtype=rasterio.float32, count=1)

        with rasterio.open(output_path, 'w', **profile) as dst:
            dst.write(ndvi.astype(rasterio.float32), 1)

    print(f"✅ NDVI 已保存到: {output_path}")


C:\Users\owner\AppData\Local\Temp\ipykernel_41332\680511140.py:36: RuntimeWarning: invalid value encountered in divide
  (nir - red) / (nir + red)


✅ NDVI 已保存到: D:\seoul\c_data\b_ndvi\b_process\ndvi_20160807.tif
✅ NDVI 已保存到: D:\seoul\c_data\b_ndvi\b_process\ndvi_20160924.tif
✅ NDVI 已保存到: D:\seoul\c_data\b_ndvi\b_process\ndvi_20230616.tif
✅ NDVI 已保存到: D:\seoul\c_data\b_ndvi\b_process\ndvi_20230819.tif


In [8]:
import os
import rasterio
import geopandas as gpd
from rasterio.mask import mask
from glob import glob

def clip_rasters_with_shapefile(raster_folder, shapefile_path, output_folder):
    # Ensure the output directory exists
    os.makedirs(output_folder, exist_ok=True)

    # Load the shapefile as a GeoDataFrame
    nyc_shape = gpd.read_file(shapefile_path)

    # Loop through each raster in the raster folder
    for raster_path in glob(os.path.join(raster_folder, '*.tif')):
        with rasterio.open(raster_path) as src:
            raster_crs = src.crs

            # Ensure the shapefile is in the same CRS as the raster
            if nyc_shape.crs != raster_crs:
                nyc_shape = nyc_shape.to_crs(raster_crs)

            # Clip the raster using the shapefile
            out_image, out_transform = mask(src, nyc_shape.geometry, crop=True)

            # Update metadata for the output raster
            out_meta = src.meta.copy()
            out_meta.update({
                "driver": "GTiff",
                "height": out_image.shape[1],
                "width": out_image.shape[2],
                "transform": out_transform,
                "crs": raster_crs  # Use the raster's CRS for the output
            })

            # Save the clipped raster to the output folder
            output_raster_path = os.path.join(output_folder, f"clipped_{os.path.basename(raster_path)}")
            if os.path.isdir(output_raster_path):
                continue
            with rasterio.open(output_raster_path, "w", **out_meta) as dest:
                dest.write(out_image)

            print(f"Clipped raster saved to: {output_raster_path}")

input_path = r'D:\seoul\c_data\b_ndvi\b_process'  # 改为文件夹路径
shp_seoul = r'D:\seoul\c_data\Admin_boundary\Seoul_boundary.shp'
output_path = r'D:\seoul\c_data\b_ndvi\c_clipped'

clip_rasters_with_shapefile(input_path, shp_seoul, output_path)

Clipped raster saved to: D:\seoul\c_data\b_ndvi\c_clipped\clipped_ndvi_20160807.tif
Clipped raster saved to: D:\seoul\c_data\b_ndvi\c_clipped\clipped_ndvi_20160924.tif
Clipped raster saved to: D:\seoul\c_data\b_ndvi\c_clipped\clipped_ndvi_20230616.tif
Clipped raster saved to: D:\seoul\c_data\b_ndvi\c_clipped\clipped_ndvi_20230819.tif


# 用 QA清理ndvi数据

In [10]:
# 用云层清理数据
import rasterio
import numpy as np
import os
import glob

extreme_heat=[20160807, 20230819]
normal_heat=[20160924, 20230616]
label_dict = {d: 'extreme' for d in extreme_heat}
label_dict.update({d: 'normal' for d in normal_heat})

for date, label in label_dict.items():
    print(f"{label} heat day:", date)
    data_workspace = rf"D:\seoul\b_satellite_img\{label}\clipped"
    ndvi_data = os.path.join(r'D:\seoul\c_data\b_ndvi\c_clipped', rf'clipped_ndvi_{date}.tif')
    qa_data = glob.glob(os.path.join(data_workspace, rf'clipped_*_{date}_*_QA_PIXEL.TIF'))
    output_folder = os.path.join(r'D:\seoul\c_data\b_ndvi', r'd_clean')
    os.makedirs(output_folder, exist_ok=True)

    ndvi_data = ndvi_data
    qa_data = qa_data[0]

    output_data = os.path.join(output_folder, rf'clipped_ndvi_{date}_cleaned.tif')
    print("NDVI Data Path:", ndvi_data)
    print("QA Data Path:", qa_data)
    print("Output Data Path:", output_data)

    # Read ndvi and QA data
    with rasterio.open(ndvi_data) as src:
        ndvi = src.read(1).astype(float)
        profile = src.profile

    with rasterio.open(qa_data) as qa_src:
        qa = qa_src.read(1)

    # QA filtering
    valid_labels = [21824, 21952]
    cloud_mask = ~np.isin(qa, valid_labels)
    ndvi[cloud_mask] = 0

    with rasterio.open(output_data, 'w', **profile) as dst:
        dst.write(ndvi, 1)


extreme heat day: 20160807
NDVI Data Path: D:\seoul\c_data\b_ndvi\c_clipped\clipped_ndvi_20160807.tif
QA Data Path: D:\seoul\b_satellite_img\extreme\clipped\clipped_LC08_L2SP_116034_20160807_20200906_02_T1_QA_PIXEL.TIF
Output Data Path: D:\seoul\c_data\b_ndvi\d_clean\clipped_ndvi_20160807_cleaned.tif
extreme heat day: 20230819
NDVI Data Path: D:\seoul\c_data\b_ndvi\c_clipped\clipped_ndvi_20230819.tif
QA Data Path: D:\seoul\b_satellite_img\extreme\clipped\clipped_LC09_L2SP_116034_20230819_20230821_02_T1_QA_PIXEL.TIF
Output Data Path: D:\seoul\c_data\b_ndvi\d_clean\clipped_ndvi_20230819_cleaned.tif
normal heat day: 20160924
NDVI Data Path: D:\seoul\c_data\b_ndvi\c_clipped\clipped_ndvi_20160924.tif
QA Data Path: D:\seoul\b_satellite_img\normal\clipped\clipped_LC08_L2SP_116034_20160924_20200906_02_T1_QA_PIXEL.TIF
Output Data Path: D:\seoul\c_data\b_ndvi\d_clean\clipped_ndvi_20160924_cleaned.tif
normal heat day: 20230616
NDVI Data Path: D:\seoul\c_data\b_ndvi\c_clipped\clipped_ndvi_20230616

In [ ]:
import rasterio
from rasterio.mask import mask
import geopandas as gpd
from shapely.geometry import mapping
import numpy as np
import os

# 输入路径
ndvi_2016_path = r'D:\seoul\c_data\b_ndvi\d_clean\clipped_ndvi_20160924_cleaned.tif'
ndvi_2023_path = r'D:\seoul\c_data\b_ndvi\d_clean\clipped_ndvi_20230616_cleaned.tif'
grid_folder = r'D:\seoul\grids\lst_map'
output_folder = grid_folder
# output_folder = r'D:\seoul\grids\lst_map\test'
os.makedirs(output_folder, exist_ok=True)

# === 循环处理所有 _bcr_bhv.shp 文件 ===
for filename in os.listdir(grid_folder):

    if filename.endswith('_bcr_bhv.shp'):
    # if filename.endswith('xy.shp'):
        input_path = os.path.join(grid_folder, filename)
        output_path = os.path.join(output_folder, filename.replace('.shp', 'ndvi.shp'))
        if filename.startswith('city2023'):
            ndvi_path = ndvi_2023_path
        elif filename.startswith('city2016'):
            ndvi_path = ndvi_2016_path
        # 读取 NDVI 影像
        with rasterio.open(ndvi_path) as src:
            ndvi_crs = src.crs

            # 读取 shapefile 并转坐标系
            grid = gpd.read_file(input_path)
            grid = grid.to_crs(ndvi_crs)

            mean_ndvi = []
            for idx, row in grid.iterrows():
                geom = [mapping(row.geometry)]
                out_image, _ = mask(src, geom, crop=True, all_touched=True)
                data = out_image[0].astype(float)
                data[data == src.nodata] = np.nan
                if np.isnan(data).all():
                    mean_val = -9999
                else:
                    mean_val = np.nanmean(data)
                mean_ndvi.append(mean_val)
            if os.path.exists(output_path):
                print(f"⏩ 已存在，跳过: {output_path}")
                pass
            # 添加 NDVI 字段并保存
            grid['NDVI'] = mean_ndvi
            original_crs = gpd.read_file(input_path).crs  # 读取原始坐标系
            grid = grid.to_crs(original_crs)  # 再转回去

            grid.to_file(output_path)
            print(f"✅ NDVI 已写入: {output_path}")


✅ NDVI 已写入: D:\seoul\grids\lst_map\test\city2016_lst_ratio_grid_1080m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp
✅ NDVI 已写入: D:\seoul\grids\lst_map\test\city2016_lst_ratio_grid_120m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp
✅ NDVI 已写入: D:\seoul\grids\lst_map\test\city2016_lst_ratio_grid_240m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp
✅ NDVI 已写入: D:\seoul\grids\lst_map\test\city2016_lst_ratio_grid_360m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp
✅ NDVI 已写入: D:\seoul\grids\lst_map\test\city2016_lst_ratio_grid_480m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp
✅ NDVI 已写入: D:\seoul\grids\lst_map\test\city2016_lst_ratio_grid_600m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp
✅ NDVI 已写入: D:\seoul\grids\lst_map\test\city2016_lst_ratio_grid_720m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp
✅ NDVI 已写入: D:\seoul\grids\lst_map\test\city2016_lst_ratio_grid_840m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp
✅ NDVI 已写入: D:\seoul\grids\lst_map\test

# 02 SVF sky view factor

In [52]:
import rasterio
from rasterio.mask import mask
import geopandas as gpd
from shapely.geometry import mapping
import numpy as np
import os

# 输入路径
svf_path = r'D:\seoul\Final_data\skyviewfactor\svf calculation_qgis\final_svf.tif'
grid_folder = r'D:\seoul\grids\lst_map'
output_folder = grid_folder
# output_folder = r'D:\seoul\grids\lst_map\test'
os.makedirs(output_folder, exist_ok=True)

# === 循环处理所有 _bcr_bhv.shp 文件 ===
for filename in os.listdir(grid_folder):
    if filename.endswith('_bcr_bhv_ndvi.shp'):
    # if filename.endswith('xy.shp'):
        input_path = os.path.join(grid_folder, filename)
        output_path = os.path.join(output_folder, filename.replace('.shp', '_svf.shp'))

        # 读取 SVF 影像
        with rasterio.open(svf_path) as src:
            svf_crs = src.crs

            # 读取 shapefile 并转坐标系
            grid = gpd.read_file(input_path)
            grid = grid.to_crs(svf_crs)
            if os.path.exists(output_path):
                print(f"⏩ 已存在，跳过: {output_path}")
                pass
            mean_svf = []
            for idx, row in grid.iterrows():
                geom = [mapping(row.geometry)]
                try:
                    out_image, _ = mask(src, geom, crop=True, all_touched=True)
                    data = out_image[0].astype(float)
                    data[data == src.nodata] = np.nan
                    mean_val = np.nanmean(data)
                    if np.isnan(mean_val):
                        mean_val = -9999
                except:
                    mean_val = np.nan
                mean_svf.append(mean_val)

            # 添加 SVF 字段并保存
            grid['SVF'] = mean_svf
            original_crs = gpd.read_file(input_path).crs  # 读取原始坐标系
            grid = grid.to_crs(original_crs)  # 再转回去

            grid.to_file(output_path)
            print(f"✅ SVF 已写入: {output_path}")


✅ SVF 已写入: D:\seoul\grids\lst_map\test\city2016_lst_ratio_grid_1080m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp
✅ SVF 已写入: D:\seoul\grids\lst_map\test\city2016_lst_ratio_grid_120m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp
✅ SVF 已写入: D:\seoul\grids\lst_map\test\city2016_lst_ratio_grid_240m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp
✅ SVF 已写入: D:\seoul\grids\lst_map\test\city2016_lst_ratio_grid_360m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp
✅ SVF 已写入: D:\seoul\grids\lst_map\test\city2016_lst_ratio_grid_480m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp
✅ SVF 已写入: D:\seoul\grids\lst_map\test\city2016_lst_ratio_grid_600m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp
✅ SVF 已写入: D:\seoul\grids\lst_map\test\city2016_lst_ratio_grid_720m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp
✅ SVF 已写入: D:\seoul\grids\lst_map\test\city2016_lst_ratio_grid_840m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp
✅ SVF 已写入: D:\seoul\grids\lst_map\test\city2016

# 02 elevation

In [7]:

import rasterio
from rasterio.mask import mask
import geopandas as gpd
from shapely.geometry import mapping
import numpy as np
import os

# 输入路径
elevation_path = r'D:\seoul\Final_data\elevation\seoul_dem.tif'
grid_folder = r'D:\seoul\grids\lst_map'
output_folder = grid_folder
os.makedirs(output_folder, exist_ok=True)

# === 循环处理所有 _bcr_bhv.shp 文件 ===
for filename in os.listdir(grid_folder):
    if filename.endswith('_bcr_bhv_ndvi_svf.shp'):
        input_path = os.path.join(grid_folder, filename)
        output_path = os.path.join(output_folder, filename.replace('.shp', '_ev.shp'))
        if os.path.exists(output_path):
            print(f"⏩ 已存在，跳过: {output_path}")
            pass
        # 读取 EV 影像
        with rasterio.open(elevation_path) as src:
            ev_crs = src.crs

            # 读取 shapefile 并转坐标系
            grid = gpd.read_file(input_path)
            grid = grid.to_crs(ev_crs)

            mean_ev = []
            for idx, row in grid.iterrows():
                geom = [mapping(row.geometry)]
                try:
                    out_image, _ = mask(src, geom, crop=True, all_touched=True)
                    data = out_image[0].astype(float)
                    data[data == src.nodata] = np.nan
                    mean_val = np.nanmean(data)
                except:
                    mean_val = np.nan
                mean_ev.append(mean_val)

            # 添加 EV 字段并保存
            grid['EV(m)'] = mean_ev
            original_crs = gpd.read_file(input_path).crs  # 读取原始坐标系
            grid = grid.to_crs(original_crs)  # 再转回去

            grid.to_file(output_path)
            print(f"✅ EV 已写入: {output_path}")


✅ EV 已写入: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_1080m_bcr_bhv_ndvi_svf_ev.shp
✅ EV 已写入: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_120m_bcr_bhv_ndvi_svf_ev.shp
✅ EV 已写入: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_240m_bcr_bhv_ndvi_svf_ev.shp
✅ EV 已写入: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_360m_bcr_bhv_ndvi_svf_ev.shp
✅ EV 已写入: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_480m_bcr_bhv_ndvi_svf_ev.shp
✅ EV 已写入: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_600m_bcr_bhv_ndvi_svf_ev.shp
✅ EV 已写入: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_720m_bcr_bhv_ndvi_svf_ev.shp
✅ EV 已写入: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_840m_bcr_bhv_ndvi_svf_ev.shp
✅ EV 已写入: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_960m_bcr_bhv_ndvi_svf_ev.shp
✅ EV 已写入: D:\seoul\grids\lst_map\city2023_lst_ratio_grid_1080m_bcr_bhv_ndvi_svf_ev.shp
✅ EV 已写入: D:\seoul\grids\lst_map\city2023_lst_ratio_grid_120m_bcr_bhv_ndvi_svf_ev.shp
✅ EV 已写入: D:\seoul\grids\lst_map\city2023_lst_ratio_

# 02 distance to park

In [9]:
# import geopandas as gpd
# import matplotlib.pyplot as plt

# # 读取 shp 文件
# shapefile_path = r'D:\seoul\Final_data\park\park_final_final.shp'
# gdf = gpd.read_file(shapefile_path)

# # 计算面积
# gdf['area_sqm'] = gdf.geometry.area  # EPSG 坐标系需为米
# gdf['area_m²'] = gdf['area_sqm']

# # 使用完整 gdf
# filtered_gdf = gdf

# # 统计数据
# max_area = filtered_gdf['area_m²'].max()
# min_area = filtered_gdf['area_m²'].min()
# mean_area = filtered_gdf['area_m²'].mean()
# median_area = filtered_gdf['area_m²'].median()
# std_area = filtered_gdf['area_m²'].std()
# count = filtered_gdf.shape[0]

# # 打印统计结果
# print("===== Seoul Parks Area 统计数据（单位：m²）=====")
# print(f"样本数量: {count}")
# print(f"最大值: {max_area:.2f} m²")
# print(f"最小值: {min_area:.2f} m²")
# print(f"平均值: {mean_area:.2f} m²")
# print(f"中位数: {median_area:.2f} m²")
# print(f"标准差: {std_area:.2f} m²")

# # 绘制直方图（横坐标单位：公顷）
# plt.figure(figsize=(10, 6))
# plt.hist(filtered_gdf['area_m²'] / 10000, bins=100, edgecolor='black')  # 换算为 hectares
# plt.title('Seoul Parks Area Distribution')
# plt.xlabel('Area (ha)')
# plt.ylabel('Count')
# plt.grid(True, linestyle='--', alpha=0.7)

# # 平均线
# plt.axvline(mean_area / 10000, color='red', linestyle='dashed', linewidth=1.5, label=f'平均: {mean_area/10000:.2f} ha')
# plt.legend()

# plt.tight_layout()
# plt.show()
# *** 01_02 替代算法，需要统计census tract中心点到绿地的最近值

import geopandas as gpd
from shapely.geometry import Point

def calculate_min_distance_to_parks(park_shp_path, census_shp_path, output_path):
    parks_gdf = gpd.read_file(park_shp_path)
    census_gdf = gpd.read_file(census_shp_path)

    # ✅ 投影一致性：转换为 UTM 等以米为单位的投影
    parks_crs = parks_gdf.crs
    census_gdf = census_gdf.to_crs(parks_crs)

    # ✅ calculate the centroid
    census_gdf = census_gdf.copy()
    census_gdf['centroid'] = census_gdf.geometry.centroid

    # ✅ calculate the distance from the point to the centroid
    distances = [parks_gdf.distance(pt).min() for pt in census_gdf['centroid']]
    census_gdf['Dist_BP'] = distances

    # ✅ 删除临时列
    census_gdf.drop(columns=['centroid'], inplace=True)

    # ✅ 坐标系换回原始
    original_crs = gpd.read_file(census_shp_path).crs
    census_gdf = census_gdf.to_crs(original_crs)

    # ✅ 保存
    census_gdf.to_file(output_path)
    print(f"✅ 已保存: {output_path}")


# 示例调用
park_shp =  r'D:\seoul\c_data\e_park\seoul_large_park.shp'
grid_folder = r'D:\seoul\grids\lst_map'
output_folder = grid_folder
os.makedirs(output_folder, exist_ok=True)

for filename in os.listdir(grid_folder):
    if filename.endswith('_bcr_bhv_ndvi_svf_ev.shp'):
        input_path = os.path.join(grid_folder, filename)
        output_path = os.path.join(output_folder, filename.replace('.shp', '_distbp.shp'))
        if os.path.exists(output_path):
            print(f"⏩ 已存在，跳过: {output_path}")
            pass
        calculate_min_distance_to_parks(park_shp, input_path, output_path)

✅ 已保存: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_1080m_bcr_bhv_ndvi_svf_ev_distbp.shp
✅ 已保存: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_120m_bcr_bhv_ndvi_svf_ev_distbp.shp
✅ 已保存: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_240m_bcr_bhv_ndvi_svf_ev_distbp.shp
✅ 已保存: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_360m_bcr_bhv_ndvi_svf_ev_distbp.shp
✅ 已保存: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_480m_bcr_bhv_ndvi_svf_ev_distbp.shp
✅ 已保存: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_600m_bcr_bhv_ndvi_svf_ev_distbp.shp
✅ 已保存: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_720m_bcr_bhv_ndvi_svf_ev_distbp.shp
✅ 已保存: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_840m_bcr_bhv_ndvi_svf_ev_distbp.shp
✅ 已保存: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_960m_bcr_bhv_ndvi_svf_ev_distbp.shp
✅ 已保存: D:\seoul\grids\lst_map\city2023_lst_ratio_grid_1080m_bcr_bhv_ndvi_svf_ev_distbp.shp
✅ 已保存: D:\seoul\grids\lst_map\city2023_lst_ratio_grid_120m_bcr_bhv_ndvi_svf_ev_distbp.shp
✅ 已保存: D

# 02 distance to mountain

In [10]:
import geopandas as gpd
from shapely.geometry import Point

def calculate_min_distance_to_mountain(park_shp_path, census_shp_path, output_path):
    parks_gdf = gpd.read_file(park_shp_path)
    census_gdf = gpd.read_file(census_shp_path)

    # ✅ 投影一致性：转换为 UTM 等以米为单位的投影
    parks_crs = parks_gdf.crs
    census_gdf = census_gdf.to_crs(parks_crs)

    # ✅ 计算中心点
    census_gdf = census_gdf.copy()
    census_gdf['centroid'] = census_gdf.geometry.centroid

    distances = [parks_gdf.distance(pt).min() for pt in census_gdf['centroid']]
    census_gdf['Dist_MT'] = distances

    # ✅ 删除临时列
    census_gdf.drop(columns=['centroid'], inplace=True)

    # ✅ 坐标系换回原始
    original_crs = gpd.read_file(census_shp_path).crs
    census_gdf = census_gdf.to_crs(original_crs)

    # ✅ 保存
    census_gdf.to_file(output_path)
    print(f"✅ 已保存: {output_path}")


# 示例调用
park_shp =  r'D:\seoul\c_data\f_mountain\mountain_final.shp'
grid_folder = r'D:\seoul\grids\lst_map'
output_folder = grid_folder
os.makedirs(output_folder, exist_ok=True)

for filename in os.listdir(grid_folder):
    if filename.endswith('_bcr_bhv_ndvi_svf_ev_distbp.shp'):
        input_path = os.path.join(grid_folder, filename)
        output_path = os.path.join(output_folder, filename.replace('.shp', '_distmt.shp'))
        if os.path.exists(output_path):
            print(f"⏩ 已存在，跳过: {output_path}")
            pass
        calculate_min_distance_to_mountain(park_shp, input_path, output_path)

✅ 已保存: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_1080m_bcr_bhv_ndvi_svf_ev_distbp_distmt.shp
✅ 已保存: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_120m_bcr_bhv_ndvi_svf_ev_distbp_distmt.shp
✅ 已保存: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_240m_bcr_bhv_ndvi_svf_ev_distbp_distmt.shp
✅ 已保存: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_360m_bcr_bhv_ndvi_svf_ev_distbp_distmt.shp
✅ 已保存: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_480m_bcr_bhv_ndvi_svf_ev_distbp_distmt.shp
✅ 已保存: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_600m_bcr_bhv_ndvi_svf_ev_distbp_distmt.shp
✅ 已保存: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_720m_bcr_bhv_ndvi_svf_ev_distbp_distmt.shp
✅ 已保存: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_840m_bcr_bhv_ndvi_svf_ev_distbp_distmt.shp
✅ 已保存: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_960m_bcr_bhv_ndvi_svf_ev_distbp_distmt.shp
✅ 已保存: D:\seoul\grids\lst_map\city2023_lst_ratio_grid_1080m_bcr_bhv_ndvi_svf_ev_distbp_distmt.shp
✅ 已保存: D:\seoul\grids\lst_ma

# 02 distance to waterbodies

In [11]:
import geopandas as gpd
from shapely.geometry import Point

def calculate_min_distance_to_waterbodies(park_shp_path, census_shp_path, output_path):
    parks_gdf = gpd.read_file(park_shp_path)
    census_gdf = gpd.read_file(census_shp_path)

    # ✅ 投影一致性：转换为 UTM 等以米为单位的投影
    parks_crs = parks_gdf.crs
    census_gdf = census_gdf.to_crs(parks_crs)

    # ✅ 计算中心点
    census_gdf = census_gdf.copy()
    census_gdf['centroid'] = census_gdf.geometry.centroid

    distances = [parks_gdf.distance(pt).min() for pt in census_gdf['centroid']]
    census_gdf['Dist_WB'] = distances

    # ✅ 删除临时列
    census_gdf.drop(columns=['centroid'], inplace=True)

    # ✅ 坐标系换回原始
    original_crs = gpd.read_file(census_shp_path).crs
    census_gdf = census_gdf.to_crs(original_crs)

    # ✅ 保存
    census_gdf.to_file(output_path)
    print(f"✅ 已保存: {output_path}")


# 示例调用
park_shp =  r'D:\seoul\c_data\g_stream\seoul_stream_final.shp'
grid_folder = r'D:\seoul\grids\lst_map'
output_folder = grid_folder
os.makedirs(output_folder, exist_ok=True)

for filename in os.listdir(grid_folder):
    if filename.endswith('_bcr_bhv_ndvi_svf_ev_distbp_distmt.shp'):
        input_path = os.path.join(grid_folder, filename)
        output_path = os.path.join(output_folder, filename.replace('.shp', '_distwb.shp'))
        if os.path.exists(output_path):
            print(f"⏩ 已存在，跳过: {output_path}")
            pass
        calculate_min_distance_to_waterbodies(park_shp, input_path, output_path)

✅ 已保存: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_1080m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb.shp
✅ 已保存: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_120m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb.shp
✅ 已保存: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_240m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb.shp
✅ 已保存: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_360m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb.shp
✅ 已保存: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_480m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb.shp
✅ 已保存: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_600m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb.shp
✅ 已保存: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_720m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb.shp
✅ 已保存: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_840m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb.shp
✅ 已保存: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_960m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb.shp
✅ 已保存: D:\seoul\grids\lst_map\city2023_lst_ratio_grid_1080m_bcr

# 02 Water ratio 

In [12]:
from shapely.geometry import Point
import geopandas as gpd
import os

def calculate_waterbody_ratio(waterbody_shp_path, grid_shp_path, output_path):
    # 读取水体 shapefile
    water_gdf = gpd.read_file(waterbody_shp_path)

    # 读取网格 shapefile
    grid_gdf = gpd.read_file(grid_shp_path)

    # 坐标系统一
    target_crs = grid_gdf.crs
    water_gdf = water_gdf.to_crs(target_crs)

    # 创建新列 WR 初始为 0
    grid_gdf['WR(%)'] = 0.0

    # 遍历每个格子，计算水体与格子的交集面积
    for idx, row in grid_gdf.iterrows():
        grid_geom = row.geometry
        intersected_water = water_gdf[water_gdf.intersects(grid_geom)]

        if not intersected_water.empty:
            clipped = gpd.overlay(intersected_water, gpd.GeoDataFrame(geometry=[grid_geom], crs=target_crs), how='intersection')
            water_area = clipped.geometry.area.sum()
            total_area = grid_geom.area
            wr_ratio = (water_area / total_area) * 100 if total_area > 0 else 0
            grid_gdf.at[idx, 'WR(%)'] = wr_ratio

    # 保存结果
    grid_gdf.to_file(output_path)
    print(f"✅ WR 已保存到: {output_path}")

waterbody_shp = r'D:\seoul\c_data\g_stream\seoul_stream_final.shp'
grid_folder = r'D:\seoul\grids\lst_map'
output_folder = grid_folder

for filename in os.listdir(grid_folder):
    if filename.endswith('_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb.shp'):
        input_path = os.path.join(grid_folder, filename)
        output_path = os.path.join(output_folder, filename.replace('.shp', '_wr.shp'))
        if os.path.exists(output_path):
            print(f"⏩ 已存在，跳过: {output_path}")
            pass
        calculate_waterbody_ratio(waterbody_shp, input_path, output_path)

✅ WR 已保存到: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_1080m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr.shp
✅ WR 已保存到: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_120m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr.shp
✅ WR 已保存到: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_240m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr.shp
✅ WR 已保存到: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_360m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr.shp
✅ WR 已保存到: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_480m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr.shp
✅ WR 已保存到: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_600m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr.shp
✅ WR 已保存到: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_720m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr.shp
✅ WR 已保存到: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_840m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr.shp
✅ WR 已保存到: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_960m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr.shp


# 增加一个代码，记录每个grid的中心点坐标

In [13]:
import os
import geopandas as gpd

grid_folder = r'D:\seoul\grids\lst_map'
output_folder = grid_folder

for filename in os.listdir(grid_folder):
    if filename.endswith('_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr.shp'):
        input_path = os.path.join(grid_folder, filename)
        output_path = os.path.join(output_folder, filename.replace('.shp', '_xy.shp'))
        if os.path.exists(output_path):
            print(f"⏩ 已存在，跳过: {output_path}")
            pass
        # 读取 shapefile
        gdf = gpd.read_file(input_path)

        # 计算每个要素的中心点坐标
        centroids = gdf.geometry.centroid
        gdf['X'] = centroids.x
        gdf['Y'] = centroids.y

        # 保存新的 shapefile
        gdf.to_file(output_path)
        print(f"✅ 中心点坐标已写入: {output_path}")

✅ 中心点坐标已写入: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_1080m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp
✅ 中心点坐标已写入: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_120m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp
✅ 中心点坐标已写入: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_240m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp
✅ 中心点坐标已写入: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_360m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp
✅ 中心点坐标已写入: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_480m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp
✅ 中心点坐标已写入: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_600m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp
✅ 中心点坐标已写入: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_720m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp
✅ 中心点坐标已写入: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_840m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp
✅ 中心点坐标已写入: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_960m_bcr_bhv_ndvi_sv

# 03 00 数据清理

In [ ]:
import os
import geopandas as gpd

grid_folder = r'D:\seoul\grids\lst_map'
output_folder = r'D:\seoul\grids\lst_map\final_clean'

os.makedirs(output_folder, exist_ok=True)

for filename in os.listdir(grid_folder):
    if filename.endswith('xy.shp'):
        filepath = os.path.join(grid_folder, filename)
        gdf = gpd.read_file(filepath)

        # 检查是否存在 -9999 的值
        if (gdf == -9999).any().any():
            # 删除包含 -9999 的行
            gdf_clean = gdf[~(gdf == -9999).any(axis=1)]

            # 保存清理后的文件
            output_path = os.path.join(output_folder, filename)
            gdf_clean.to_file(output_path)

            print(f"✔ 已清理并保存：{filename}，原始行数 {len(gdf)}, 清理后行数 {len(gdf_clean)}")
        else:
            print(f"✅ 无需清理：{filename}")


✔ 已清理并保存：city2016_lst_ratio_grid_1080m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp，原始行数 429, 清理后行数 416
✔ 已清理并保存：city2016_lst_ratio_grid_120m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp，原始行数 41088, 清理后行数 35294
✔ 已清理并保存：city2016_lst_ratio_grid_240m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp，原始行数 10049, 清理后行数 8953
✔ 已清理并保存：city2016_lst_ratio_grid_360m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp，原始行数 4379, 清理后行数 3993
✔ 已清理并保存：city2016_lst_ratio_grid_480m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp，原始行数 2407, 清理后行数 2245
✔ 已清理并保存：city2016_lst_ratio_grid_600m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp，原始行数 1514, 清理后行数 1423
✔ 已清理并保存：city2016_lst_ratio_grid_720m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp，原始行数 1029, 清理后行数 981
✔ 已清理并保存：city2016_lst_ratio_grid_840m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp，原始行数 736, 清理后行数 710
✔ 已清理并保存：city2016_lst_ratio_grid_960m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp，原始行数 557, 清理后行数 539
✔ 已清

# 03 出图！

In [25]:
import sys
! {sys.executable} -m pip install mapclassify>=2.4.0


[notice] A new release of pip is available: 24.3.1 -> 25.1.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [56]:

import geopandas as gpd
import matplotlib.pyplot as plt
import matplotlib.patches as mpatches
import os
import numpy as np
# from mapclassify import Quantiles
from mapclassify import NaturalBreaks
# 输出文件夹
output_folder = r'D:\seoul\grids\lst_map\figures\grids_colored'
os.makedirs(output_folder, exist_ok=True)
seoul_boundary_file = r'D:\seoul\c_data\Admin_boundary\Seoul_boundary.shp'
shp_seoul = seoul_boundary_file
# 可视化字段
explanatory_vars = ['BCR(%)', 'BHA','BHV', 'NDVI', 'SVF', 'EV(m)',
                    'Dist_BP', 'Dist_MT', 'Dist_WB', 'WR(%)']
grid_sizes = range(120, 1200, 120)

for value_column in explanatory_vars:
    for i in grid_sizes:
        for year in [2016, 2023]:
            shp_path = rf'D:\seoul\grids\lst_map\final_clean\city{year}_lst_ratio_grid_{i}m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp'

            if not os.path.exists(shp_path):
                print(f"❌ 文件不存在：{shp_path}")
                continue

            gdf = gpd.read_file(shp_path)
            boundary = gpd.read_file(shp_seoul)
            if value_column not in gdf.columns:
                print(f"⚠️ {value_column} 不存在于 {i}m 数据中，跳过")
                continue

            # 创建分级分类器
            classifier = NaturalBreaks(gdf[value_column], k=5)
            bins = classifier.bins
            gdf['class'] = classifier.yb  # 用于分类着色

            # 色带
            cmap = plt.get_cmap('RdYlBu_r')
            colors = cmap(np.linspace(0, 1, 5))
            # 统一坐标参考系
            if boundary.crs != gdf.crs:
                boundary = boundary.to_crs(gdf.crs)

            # 开始绘图
            fig, ax = plt.subplots(figsize=(9, 8))

            # 先绘制首尔边界
            boundary.plot(ax=ax, color='white', edgecolor='black', linewidth=1.0, zorder=0)

            # 再绘制分级网格
            for cls in range(5):
                gdf[gdf['class'] == cls].plot(ax=ax, color=colors[cls], linewidth=0.2, edgecolor='gray', zorder=1)

            # 设置标题
            ax.set_title(f'{year} {value_column} (Grid: {i}m)', fontsize=14)
            ax.axis('off')

            # 构建矩形 legend
            patches = []
            for j in range(5):
                if j == 0:
                    label = f'≤ {bins[j]:.2f}'
                else:
                    label = f'{bins[j-1]:.2f} - {bins[j]:.2f}'
                patch = mpatches.Patch(color=colors[j], label=label)
                patches.append(patch)

            # 添加图例到右边
            ax.legend(handles=patches,
                      title=value_column,  # ✅ 图例标题
                      loc='upper left',
                      bbox_to_anchor=(0.85, 0.95),
                      frameon=True)

            # 保存图像
            output_path = os.path.join(output_folder, f'{year}_{value_column}_{i}m_map.png')
            fig.subplots_adjust(left=0.05, right=0.85, top=0.95, bottom=0.05)  # ❷ 保证图例不会影响主图大小
            plt.savefig(output_path, dpi=300)

            plt.close()
            print(f"✅ Saved: {output_path}")


✅ Saved: D:\seoul\grids\lst_map\figures\grids_colored\2016_BCR(%)_120m_map.png
✅ Saved: D:\seoul\grids\lst_map\figures\grids_colored\2023_BCR(%)_120m_map.png
✅ Saved: D:\seoul\grids\lst_map\figures\grids_colored\2016_BCR(%)_240m_map.png
✅ Saved: D:\seoul\grids\lst_map\figures\grids_colored\2023_BCR(%)_240m_map.png
✅ Saved: D:\seoul\grids\lst_map\figures\grids_colored\2016_BCR(%)_360m_map.png
✅ Saved: D:\seoul\grids\lst_map\figures\grids_colored\2023_BCR(%)_360m_map.png
✅ Saved: D:\seoul\grids\lst_map\figures\grids_colored\2016_BCR(%)_480m_map.png
✅ Saved: D:\seoul\grids\lst_map\figures\grids_colored\2023_BCR(%)_480m_map.png
✅ Saved: D:\seoul\grids\lst_map\figures\grids_colored\2016_BCR(%)_600m_map.png
✅ Saved: D:\seoul\grids\lst_map\figures\grids_colored\2023_BCR(%)_600m_map.png
✅ Saved: D:\seoul\grids\lst_map\figures\grids_colored\2016_BCR(%)_720m_map.png
✅ Saved: D:\seoul\grids\lst_map\figures\grids_colored\2023_BCR(%)_720m_map.png
✅ Saved: D:\seoul\grids\lst_map\figures\grids_colore

In [ ]:
# import geopandas as gpd
# import os

# input_folder = r'D:\seoul\grids\lst_map'
# explanatory_vars = ['BCR(%)', 'BHA','BHV', 'NDVI', 'SVF', 'EV(m)',
#                     'Dist_BP', 'Dist_MT', 'Dist_WB', 'WR(%)']
# grid_sizes = range(120, 1200, 120)
# years = [2016, 2023]

# for value_column in explanatory_vars:
#     for i in grid_sizes:
#         for year in years:
#             shp_path = rf'D:\seoul\grids\lst_map\final_clean\city{year}_lst_ratio_grid_{i}m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp'
#             if not os.path.exists(shp_path):
#                 continue

#             gdf = gpd.read_file(shp_path)

#             if value_column not in gdf.columns:
#                 continue

#             # 检查该字段是否有NaN
#             if gdf[value_column].isna().any():
#                 nan_count = gdf[value_column].isna().sum()
#                 print(f"⚠️ 文件: {shp_path} 中字段 '{value_column}' 有 {nan_count} 个 NaN 值")


⚠️ 文件: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_120m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp 中字段 'NDVI' 有 107 个 NaN 值
⚠️ 文件: D:\seoul\grids\lst_map\city2023_lst_ratio_grid_120m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp 中字段 'NDVI' 有 119 个 NaN 值
⚠️ 文件: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_240m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp 中字段 'NDVI' 有 10 个 NaN 值
⚠️ 文件: D:\seoul\grids\lst_map\city2023_lst_ratio_grid_240m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp 中字段 'NDVI' 有 8 个 NaN 值
⚠️ 文件: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_360m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp 中字段 'NDVI' 有 3 个 NaN 值
⚠️ 文件: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_480m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp 中字段 'NDVI' 有 1 个 NaN 值
⚠️ 文件: D:\seoul\grids\lst_map\city2016_lst_ratio_grid_120m_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp 中字段 'SVF' 有 2 个 NaN 值
⚠️ 文件: D:\seoul\grids\lst_map\city2023_lst_ratio_grid_120m_bcr_bh

# 03 Descriptive statistics

In [60]:
import os
import re
import pandas as pd
import geopandas as gpd
grid_folder = r'D:\seoul\grids\lst_map\final_clean'
output_folder = os.path.join(grid_folder,r'statistics')
os.makedirs(output_folder, exist_ok=True)

for filename in os.listdir(grid_folder):
    if filename.endswith('_bcr_bhv_ndvi_svf_ev_distbp_distmt_distwb_wr_xy.shp'):
        input_path = os.path.join(grid_folder, filename)
        data = gpd.read_file(input_path)
        #print(data)
        # print("列名如下：")
        # print(list(data.columns))

for year in [2016,2023]:
    output_path = os.path.join(output_folder,f'{year} Descriptive statistics.xlsx')
    all_stats = []
    for filename in os.listdir(grid_folder):
        if filename.endswith('_xy.shp') and filename.startswith(f'city{year}'):
            input_path = os.path.join(grid_folder, filename)
            df = gpd.read_file(input_path)

            raw_columns = [f'nor_{year}', f'ext_{year}', f'hr_{year}','BHA' ,'BHV','BCR(%)', 'SVF', 'NDVI', 'EV(m)', 'WR(%)', 'Dist_BP', 'Dist_WB', 'Dist_MT']
            output_labels = ['Regular_LST(℃)', 'Extreme_LST(℃)', 'HR_T(℃)', 'BHA' ,'BHV', 'BCR(%)', 'SVF', 'NDVI', 'EV(m)', 'WR(%)', 'Near_P(m)', 'Near_S(m)', 'Near_M(m)']
            desc = df[raw_columns].describe().T[['mean', 'min', 'max', 'std']]
            desc.columns = ['Mean', 'Min', 'Max', 'SD']

            desc.index = output_labels
            desc = desc.round(2)

            # ✅ 用正则表达式提取 "{grid}"m
            match = re.search(r'(\d{3,5})m', filename)
            grid_id = match.group(1) if match else 'Unknown'

            desc['Grid'] = grid_id 
            all_stats.append(desc.reset_index())

    # 合并所有结果
    summary_df = pd.concat(all_stats, ignore_index=True)
    summary_df.rename(columns={'index': 'Variable'}, inplace=True)

    # 输出 Excel
    summary_df.to_excel(output_path, index=False)
    print(f"📁 描述性统计合并完成！保存路径：{output_path}")

📁 描述性统计合并完成！保存路径：D:\seoul\grids\lst_map\final_clean\statistics\2016 Descriptive statistics.xlsx
📁 描述性统计合并完成！保存路径：D:\seoul\grids\lst_map\final_clean\statistics\2023 Descriptive statistics.xlsx
